// @openclaw/qqbot/dist/channel-D1UztsnG.js
import { t as asOptionalObjectRecord } from "./string-normalize-R_0cKO7Q.js";
import { a as qqbotSetupAdapterShared, i as qqbotMeta, n as qqbotSetupWizard, o as ensurePlatformAdapter, r as qqbotConfigAdapter, s as getBridgeLogger, t as qqbotChannelConfigSchema } from "./config-schema-JZEf1dvB.js";
import { a as resolveQQBotAccount, n as applyQQBotAccountConfig, r as listQQBotAccountIds, t as DEFAULT_ACCOUNT_ID$1 } from "./config-C1qZbh0K.js";
import { t as getQQBotRuntime } from "./runtime-DodcT_fQ.js";
import { n as legacyConfigRules, r as normalizeCompatibilityConfig, t as buildQQBotStateKey } from "./state-keys-CQKlAFyo.js";
import { n as normalizeTarget, r as parseTarget, t as looksLikeQQBotTarget } from "./target-parser-DagYRQkP.js";
import { getExecApprovalReplyMetadata } from "openclaw/plugin-sdk/approval-runtime";
import { buildChannelOutboundSessionRoute } from "openclaw/plugin-sdk/channel-core";
import { createMessageReceiptFromOutboundResults, defineChannelMessageAdapter } from "openclaw/plugin-sdk/channel-outbound";
import { createLazyRuntimeModule } from "openclaw/plugin-sdk/lazy-runtime";
import { sanitizeAssistantVisibleText } from "openclaw/plugin-sdk/text-chunking";
import { createChannelApprovalCapability } from "openclaw/plugin-sdk/approval-delivery-runtime";
import { createLazyChannelApprovalNativeRuntimeAdapter } from "openclaw/plugin-sdk/approval-handler-adapter-runtime";
import { resolveApprovalRequestChannelAccountId, resolveApprovalRequestSessionConversation } from "openclaw/plugin-sdk/approval-native-runtime";
import { normalizeLowercaseStringOrEmpty, normalizeOptionalString } from "openclaw/plugin-sdk/string-coerce-runtime";
import { truncateUtf16Safe } from "openclaw/plugin-sdk/text-utility-runtime";
import { markImplicitSameChatApprovalAuthorization, resolveApprovalApprovers } from "openclaw/plugin-sdk/approval-auth-runtime";
import { createChannelExecApprovalProfile, isChannelExecApprovalClientEnabledFromConfig, matchesApprovalRequestFilters } from "openclaw/plugin-sdk/approval-client-runtime";
import { normalizeAccountId } from "openclaw/plugin-sdk/routing";
import { resolveChannelGroupToolsPolicy } from "openclaw/plugin-sdk/channel-policy";
//#region extensions/qqbot/src/engine/approval/index.ts
/**
* Approval helpers — pure functions, zero framework dependencies.
*
* - Build approval message text + inline keyboard
* - Resolve delivery target from session metadata
* - Parse INTERACTION_CREATE button data
*/
function buildExecApprovalText(request) {
	const expiresIn = Math.max(0, Math.round((request.expiresAtMs - Date.now()) / 1e3));
	const lines = ["🔐 命令执行审批", ""];
	const cmd = request.request.commandPreview ?? request.request.command ?? "";
	if (cmd) lines.push(`\`\`\`\n${truncateUtf16Safe(cmd, 300)}\n\`\`\``);
	if (request.request.cwd) lines.push(`\u{1f4c1} \u76ee\u5f55: ${request.request.cwd}`);
	if (request.request.agentId) lines.push(`\u{1f916} Agent: ${request.request.agentId}`);
	lines.push("", `\u23f1\ufe0f \u8d85\u65f6: ${expiresIn} \u79d2`);
	return lines.join("\n");
}
function buildPluginApprovalText(request) {
	const timeoutSec = Math.round((request.request.timeoutMs ?? 12e4) / 1e3);
	const lines = [`${request.request.severity === "critical" ? "🔴" : request.request.severity === "info" ? "🔵" : "🟡"} \u5ba1\u6279\u8bf7\u6c42`, ""];
	lines.push(`\u{1f4cb} ${request.request.title}`);
	if (request.request.description) lines.push(`\u{1f4dd} ${request.request.description}`);
	if (request.request.toolName) lines.push(`\u{1f527} \u5de5\u5177: ${request.request.toolName}`);
	if (request.request.pluginId) lines.push(`\u{1f50c} \u63d2\u4ef6: ${request.request.pluginId}`);
	if (request.request.agentId) lines.push(`\u{1f916} Agent: ${request.request.agentId}`);
	lines.push("", `\u23f1\ufe0f \u8d85\u65f6: ${timeoutSec} \u79d2`);
	return lines.join("\n");
}
/**
* Build the three-button inline keyboard for approval messages.
*
* type=1 (Callback): click triggers INTERACTION_CREATE, button_data = data field.
* group_id "approval": clicking one button grays out the others (mutual exclusion).
* click_limit=1: each user can only click once.
* permission.type=2: all users can interact.
*/
function buildApprovalKeyboard(approvalId, allowedDecisions = [
	"allow-once",
	"allow-always",
	"deny"
]) {
	const makeBtn = (id, label, visitedLabel, data, style) => ({
		id,
		render_data: {
			label,
			visited_label: visitedLabel,
			style
		},
		action: {
			type: 1,
			data,
			permission: { type: 2 },
			click_limit: 1
		},
		group_id: "approval"
	});
	const buttons = [];
	if (allowedDecisions.includes("allow-once")) buttons.push(makeBtn("allow", "✅ 允许一次", "已允许", `approve:${approvalId}:allow-once`, 1));
	if (allowedDecisions.includes("allow-always")) buttons.push(makeBtn("always", "⭐ 始终允许", "已始终允许", `approve:${approvalId}:allow-always`, 1));
	if (allowedDecisions.includes("deny")) buttons.push(makeBtn("deny", "❌ 拒绝", "已拒绝", `approve:${approvalId}:deny`, 0));
	return { content: { rows: [{ buttons }] } };
}
/**
* Extract the delivery target from a sessionKey or turnSourceTo string.
*
* Expected formats:
*   agent:main:qqbot:direct:OPENID  -> { type: "c2c", id: "OPENID" }
*   agent:main:qqbot:c2c:OPENID     -> { type: "c2c", id: "OPENID" }
*   agent:main:qqbot:group:GROUPID  -> { type: "group", id: "GROUPID" }
*
* Returns null if neither field matches the expected pattern.
*/
function resolveApprovalTarget(sessionKey, turnSourceTo) {
	const sk = sessionKey ?? turnSourceTo;
	if (!sk) return null;
	const m = sk.match(/qqbot:(c2c|direct|group):([A-F0-9]+)/i);
	if (!m) return null;
	return {
		type: m[1].toLowerCase() === "group" ? "group" : "c2c",
		id: m[2]
	};
}
/**
* Parse the button_data string from an INTERACTION_CREATE event.
*
* Expected format: `approve:<approvalId>:<decision>`
* where approvalId may be prefixed with "exec:" or "plugin:".
*
* Returns null if the data does not match the approval button format.
*/
function parseApprovalButtonData(buttonData) {
	const m = buttonData.match(/^approve:((?:(?:exec|plugin):)?[0-9a-f-]+):(allow-once|allow-always|deny)$/i);
	if (!m) return null;
	return {
		approvalId: m[1],
		decision: m[2]
	};
}
//#endregion
//#region extensions/qqbot/src/exec-approvals.ts
function normalizeApproverId(value) {
	return normalizeOptionalString(String(value)) || void 0;
}
function resolveQQBotExecApprovalConfig(params) {
	const account = resolveQQBotAccount(params.cfg, params.accountId);
	const config = account.config.execApprovals;
	if (!config) return;
	return {
		...config,
		enabled: account.enabled && account.secretSource !== "none" ? config.enabled : false
	};
}
function getQQBotExecApprovalApprovers(params) {
	const accountConfig = resolveQQBotAccount(params.cfg, params.accountId).config;
	return resolveApprovalApprovers({
		explicit: resolveQQBotExecApprovalConfig(params)?.approvers,
		allowFrom: accountConfig.allowFrom,
		normalizeApprover: normalizeApproverId
	});
}
function countQQBotExecApprovalEligibleAccounts(params) {
	return listQQBotAccountIds(params.cfg).filter((accountId) => {
		const account = resolveQQBotAccount(params.cfg, accountId);
		if (!account.enabled || account.secretSource === "none") return false;
		const config = resolveQQBotExecApprovalConfig({
			cfg: params.cfg,
			accountId
		});
		return isChannelExecApprovalClientEnabledFromConfig({
			enabled: config?.enabled,
			approverCount: getQQBotExecApprovalApprovers({
				cfg: params.cfg,
				accountId
			}).length
		}) && matchesApprovalRequestFilters({
			request: params.request.request,
			agentFilter: config?.agentFilter,
			sessionFilter: config?.sessionFilter,
			fallbackAgentIdFromSessionKey: true
		});
	}).length;
}
function matchesQQBotRequestAccount(params) {
	const turnSourceChannel = normalizeLowercaseStringOrEmpty(params.request.request.turnSourceChannel);
	const boundAccountId = resolveApprovalRequestChannelAccountId({
		cfg: params.cfg,
		request: params.request,
		channel: "qqbot"
	});
	if (turnSourceChannel && turnSourceChannel !== "qqbot" && !boundAccountId) return countQQBotExecApprovalEligibleAccounts({
		cfg: params.cfg,
		request: params.request
	}) <= 1;
	return !boundAccountId || !params.accountId || normalizeAccountId(boundAccountId) === normalizeAccountId(params.accountId);
}
/**
* Count QQBot accounts that could actually deliver a native approval
* message — i.e. accounts that are enabled and have resolvable secrets.
* Disabled or unconfigured accounts never spawn a handler, so they
* must not contribute to the single-account shortcut in the fallback
* ownership check below.
*/
function countQQBotFallbackEligibleAccounts(cfg) {
	return listQQBotAccountIds(cfg).filter((accountId) => {
		const account = resolveQQBotAccount(cfg, accountId);
		return account.enabled && account.secretSource !== "none";
	}).length;
}
/**
* Fallback account-ownership check — applied when `execApprovals` is NOT
* configured for any QQBot account. In this mode every enabled account
* handler would otherwise race to deliver the same approval to its own
* openid namespace, so we must enforce per-account isolation.
*
* Rules:
*   - If the request carries a bound account (via `turnSourceAccountId`
*     or session binding), only the handler whose `accountId` matches it
*     delivers the approval. This is strict: a handler with an unknown
*     `accountId` (null/undefined) must not claim a bound request.
*   - If no account is bound, only deliver when there is a single
*     *eligible* QQBot account (enabled + secret resolved). Disabled or
*     unconfigured accounts never deliver anyway, so they shouldn't
*     block the remaining single account from handling the approval.
*     Multiple eligible accounts cannot safely race because openids are
*     account-scoped — cross-account delivery hits the QQ Bot API with
*     a mismatched token and fails.
*/
function matchesQQBotFallbackRequestAccount(params) {
	const boundAccountId = resolveApprovalRequestChannelAccountId({
		cfg: params.cfg,
		request: params.request,
		channel: "qqbot"
	});
	if (boundAccountId) {
		if (!params.accountId) return false;
		return normalizeAccountId(boundAccountId) === normalizeAccountId(params.accountId);
	}
	return countQQBotFallbackEligibleAccounts(params.cfg) <= 1;
}
/**
* Unified per-account ownership check used by both the profile and
* fallback approval paths. Dispatches to the profile rules when the
* current account has `execApprovals` configured, otherwise uses the
* fallback rules.
*
* This is the single source of truth for "does this QQBot handler own
* this approval request?" and is consumed by both the capability
* gate (shouldHandle) and the lazy native runtime adapter.
*/
function matchesQQBotApprovalAccount(params) {
	const normalized = {
		cfg: params.cfg,
		accountId: params.accountId,
		request: params.request
	};
	if (resolveQQBotExecApprovalConfig(normalized) !== void 0) return matchesQQBotRequestAccount(normalized);
	return matchesQQBotFallbackRequestAccount(normalized);
}
const qqbotExecApprovalProfile = createChannelExecApprovalProfile({
	resolveConfig: resolveQQBotExecApprovalConfig,
	resolveApprovers: getQQBotExecApprovalApprovers,
	matchesRequestAccount: matchesQQBotRequestAccount,
	fallbackAgentIdFromSessionKey: true,
	requireClientEnabledForLocalPromptSuppression: false
});
const isQQBotExecApprovalClientEnabled = qqbotExecApprovalProfile.isClientEnabled;
const isQQBotExecApprovalApprover = qqbotExecApprovalProfile.isApprover;
const isQQBotExecApprovalAuthorizedSender = qqbotExecApprovalProfile.isAuthorizedSender;
const shouldHandleQQBotExecApprovalRequest = qqbotExecApprovalProfile.shouldHandleRequest;
function authorizeQQBotApprovalAction(params) {
	if (resolveQQBotExecApprovalConfig(params) === void 0) return markImplicitSameChatApprovalAuthorization({ authorized: true });
	return (params.approvalKind === "plugin" ? isQQBotExecApprovalApprover(params) : isQQBotExecApprovalAuthorizedSender(params)) ? { authorized: true } : {
		authorized: false,
		reason: "You are not authorized to approve this request."
	};
}
//#endregion
//#region extensions/qqbot/src/bridge/approval/capability.ts
/**
* QQ Bot Approval Capability — entry point.
*
* QQBot uses a simpler approval model than Telegram/Slack: when no
* approver list is configured, the bot sends the approval message to the
* originating conversation and any participant can approve from there.
*
* When `execApprovals` IS configured, it gates which requests are
* handled natively and who is authorized.  When it is NOT configured,
* QQBot falls back to "always handle, anyone can approve".
*/
/**
* When `execApprovals` is configured, delegate to the profile-based
* check.  Otherwise fall back to target-resolvability plus the shared
* per-account ownership rule in `matchesQQBotApprovalAccount` so that
* each QQBot account handler only delivers approvals that originated
* from its own account (openids are account-scoped — cross-account
* delivery fails with 500 on the QQ Bot API).
*/
function shouldHandleRequest(params) {
	if (hasExecApprovalConfig(params)) return shouldHandleQQBotExecApprovalRequest(params);
	if (!canResolveTarget(params.request)) return false;
	return matchesQQBotApprovalAccount({
		cfg: params.cfg,
		accountId: params.accountId,
		request: params.request
	});
}
function hasExecApprovalConfig(params) {
	return resolveQQBotExecApprovalConfig(params) !== void 0;
}
function isNativeDeliveryEnabled(params) {
	if (hasExecApprovalConfig(params)) return isQQBotExecApprovalClientEnabled(params);
	const account = resolveQQBotAccount(params.cfg, params.accountId);
	return account.enabled && account.secretSource !== "none";
}
function canResolveTarget(request) {
	if (resolveApprovalTarget(request.request.sessionKey ?? null, request.request.turnSourceTo ?? null)) return true;
	return resolveApprovalRequestSessionConversation({
		request,
		channel: "qqbot",
		bundledFallback: true
	})?.id != null;
}
function resolveNativeDeliveryState(params) {
	return isNativeDeliveryEnabled(params) ? { kind: "enabled" } : { kind: "disabled" };
}
function createQQBotApprovalCapability() {
	return createChannelApprovalCapability({
		authorizeActorAction: ({ cfg, accountId, senderId, approvalKind }) => authorizeQQBotApprovalAction({
			cfg,
			accountId,
			senderId,
			approvalKind
		}),
		getActionAvailabilityState: resolveNativeDeliveryState,
		getExecInitiatingSurfaceState: resolveNativeDeliveryState,
		describeExecApprovalSetup: ({ accountId }) => {
			return `QQBot native exec approvals are enabled by default. To restrict who can approve, configure \`${accountId && accountId !== "default" ? `channels.qqbot.accounts.${accountId}` : "channels.qqbot"}.execApprovals.approvers\` with QQ user OpenIDs.`;
		},
		delivery: {
			hasConfiguredDmRoute: () => true,
			shouldSuppressForwardingFallback: (input) => {
				const channel = normalizeOptionalString(input.target?.channel);
				if (channel !== "qqbot") return false;
				const accountId = normalizeOptionalString(input.target?.accountId) ?? normalizeOptionalString(input.request?.request?.turnSourceAccountId);
				const result = isNativeDeliveryEnabled({
					cfg: input.cfg,
					accountId
				});
				getBridgeLogger().debug?.(`[qqbot:approval] shouldSuppressForwardingFallback channel=${channel} accountId=${accountId} → ${result}`);
				return result;
			}
		},
		native: {
			describeDeliveryCapabilities: ({ cfg, accountId }) => ({
				enabled: isNativeDeliveryEnabled({
					cfg,
					accountId
				}),
				preferredSurface: "origin",
				supportsOriginSurface: true,
				supportsApproverDmSurface: false,
				notifyOriginWhenDmOnly: false
			}),
			resolveOriginTarget: ({ request }) => {
				const target = resolveApprovalTarget(request.request.sessionKey ?? null, request.request.turnSourceTo ?? null);
				if (target) return { to: `${target.type}:${target.id}` };
				const sessionConversation = resolveApprovalRequestSessionConversation({
					request,
					channel: "qqbot",
					bundledFallback: true
				});
				if (sessionConversation?.id) return { to: `${sessionConversation.kind === "group" ? "group" : "c2c"}:${sessionConversation.id}` };
				return null;
			}
		},
		nativeRuntime: createLazyChannelApprovalNativeRuntimeAdapter({
			eventKinds: ["exec", "plugin"],
			isConfigured: ({ cfg, accountId }) => {
				const result = isNativeDeliveryEnabled({
					cfg,
					accountId
				});
				getBridgeLogger().debug?.(`[qqbot:approval] nativeRuntime.isConfigured accountId=${accountId} → ${result}`);
				return result;
			},
			shouldHandle: ({ cfg, accountId, request }) => {
				const result = shouldHandleRequest({
					cfg,
					accountId,
					request
				});
				getBridgeLogger().debug?.(`[qqbot:approval] nativeRuntime.shouldHandle accountId=${accountId} → ${result}`);
				return result;
			},
			load: async () => {
				ensurePlatformAdapter();
				return (await import("./handler-runtime-zQvT6SrI.js")).qqbotApprovalNativeRuntime;
			}
		})
	});
}
const qqbotApprovalCapability = createQQBotApprovalCapability();
let cachedCapability;
function getQQBotApprovalCapability() {
	cachedCapability ??= qqbotApprovalCapability;
	return cachedCapability;
}
//#endregion
//#region extensions/qqbot/src/bridge/narrowing.ts
/**
* Map resolved plugin account to the engine gateway account shape (single assertion on nested config).
*/
function toGatewayAccount(account) {
	return {
		accountId: account.accountId,
		appId: account.appId,
		clientSecret: account.clientSecret,
		markdownSupport: account.markdownSupport,
		systemPrompt: account.systemPrompt,
		config: account.config
	};
}
/**
* Persist OpenClaw config through the injected plugin runtime (typed entry point).
*/
async function writeOpenClawConfigThroughRuntime(runtime, cfg) {
	await runtime.config.replaceConfigFile({
		nextConfig: cfg,
		afterWrite: { mode: "auto" }
	});
}
//#endregion
//#region extensions/qqbot/src/doctor.ts
const qqbotDoctor = {
	legacyConfigRules,
	normalizeCompatibilityConfig
};
//#endregion
//#region extensions/qqbot/src/engine/utils/sqlite-state.ts
function resolveStoreEnv(options) {
	if (!options.stateDir) return options.env;
	return {
		...options.env ?? process.env,
		OPENCLAW_STATE_DIR: options.stateDir
	};
}
function openQQBotSyncKeyedStore(options) {
	return getQQBotRuntime().state.openSyncKeyedStore({
		namespace: options.namespace,
		maxEntries: options.maxEntries,
		...options.defaultTtlMs != null ? { defaultTtlMs: options.defaultTtlMs } : {},
		...resolveStoreEnv(options) ? { env: resolveStoreEnv(options) } : {}
	});
}
//#endregion
//#region extensions/qqbot/src/engine/config/credential-backup.ts
/**
* Credential backup & recovery.
* 凭证暂存与恢复。
*
* Solves the "hot-upgrade interrupted, appId/secret vanished from
* openclaw.json" failure mode.
*
* Mechanics:
*   - After each successful gateway start we snapshot the currently
*     resolved `appId` / `clientSecret` to a per-account SQLite KV entry.
*   - During plugin startup, if the live config has an empty appId or
*     secret, the gateway consults the backup and restores the values
*     via the config mutation API.
*   - Legacy JSON backups are imported by `openclaw doctor --fix`, not by
*     runtime startup.
*
* Safety notes:
*   - Only restore when credentials are **actually empty** — never
*     overwrite a user's intentional config change.
*   - Per-account key only; not keyed by appId because recovery happens
*     precisely when appId is unknown.
*/
const CREDENTIAL_BACKUPS_NAMESPACE = "credential-backups";
const MAX_CREDENTIAL_BACKUPS = 1e3;
function createCredentialBackupStore() {
	return openQQBotSyncKeyedStore({
		namespace: CREDENTIAL_BACKUPS_NAMESPACE,
		maxEntries: MAX_CREDENTIAL_BACKUPS
	});
}
function credentialBackupKey(accountId) {
	return buildQQBotStateKey("credential-backup", accountId);
}
function isUsableBackup(data) {
	return Boolean(data?.accountId && data.appId && data.clientSecret);
}
/** Persist a credential snapshot (called once gateway reaches READY). */
function saveCredentialBackup(accountId, appId, clientSecret) {
	if (!appId || !clientSecret) return;
	try {
		const data = {
			accountId,
			appId,
			clientSecret,
			savedAt: (/* @__PURE__ */ new Date()).toISOString()
		};
		createCredentialBackupStore().register(credentialBackupKey(accountId), data);
	} catch {}
}
/**
* Load a credential snapshot for `accountId`.
*
* Reads SQLite only. Legacy JSON backup import is owned by doctor/setup
* migration so runtime startup stays canonical-state-only.
*/
function loadCredentialBackup(accountId) {
	try {
		if (accountId) {
			const data = createCredentialBackupStore().lookup(credentialBackupKey(accountId));
			if (isUsableBackup(data)) return data;
		}
	} catch {}
	return null;
}
//#endregion
//#region extensions/qqbot/src/engine/config/credentials.ts
/**
* QQBot credential management (pure logic layer).
* QQBot 凭证管理（纯逻辑层）。
*
* Credential clearing and field-level cleanup for logout and setup
* flows. All functions operate on plain objects (Record<string, unknown>)
* and stay framework-agnostic.
*/
/**
* Remove clientSecret / clientSecretFile from a QQBot account config.
*
* Returns a shallow-cloned config with credentials removed, plus flags
* indicating whether anything actually changed.
*/
function clearAccountCredentials(cfg, accountId) {
	const nextCfg = { ...cfg };
	const channels = asOptionalObjectRecord(cfg.channels);
	const nextQQBot = channels?.qqbot ? { ...asOptionalObjectRecord(channels.qqbot) } : void 0;
	let cleared = false;
	let changed = false;
	if (nextQQBot) {
		const qqbot = nextQQBot;
		if (accountId === "default") {
			if (qqbot.clientSecret) {
				delete qqbot.clientSecret;
				cleared = true;
				changed = true;
			}
			if (qqbot.clientSecretFile) {
				delete qqbot.clientSecretFile;
				cleared = true;
				changed = true;
			}
		}
		const accounts = qqbot.accounts;
		if (accounts && accountId in accounts) {
			const entry = accounts[accountId];
			if (entry && "clientSecret" in entry) {
				delete entry.clientSecret;
				cleared = true;
				changed = true;
			}
			if (entry && "clientSecretFile" in entry) {
				delete entry.clientSecretFile;
				cleared = true;
				changed = true;
			}
			if (entry && Object.keys(entry).length === 0) {
				delete accounts[accountId];
				changed = true;
			}
		}
	}
	if (changed && nextQQBot) nextCfg.channels = {
		...channels,
		qqbot: nextQQBot
	};
	return {
		nextCfg,
		cleared,
		changed
	};
}
//#endregion
//#region extensions/qqbot/src/engine/messaging/markdown-table-chunking.ts
const QQBOT_MARKDOWN_SAFE_CHUNK_BYTE_LIMIT = 3600;
function chunkQQBotMarkdownText(text, limit, baseChunker) {
	const chunker = createQQBotMarkdownChunker(baseChunker);
	return [...chunker.chunkText(text, limit), ...chunker.flushPendingText(limit)];
}
function createQQBotMarkdownChunker(baseChunker) {
	const state = new QQBotMarkdownChunkingState(baseChunker);
	return {
		chunkText: (text, limit) => state.chunkText(text, limit),
		flushPendingText: (limit) => state.flushPendingText(limit)
	};
}
var QQBotMarkdownChunkingState = class {
	constructor(baseChunker) {
		this.baseChunker = baseChunker;
		this.activeTable = null;
		this.pendingHeaderLine = null;
		this.pendingHeaderCells = [];
		this.tableLines = [];
		this.textLines = [];
		this.pendingRowFragment = null;
		this.activeFence = null;
		this.pendingTextFenceOpenLine = null;
		this.pendingFenceLineFragment = null;
	}
	chunkText(text, limit) {
		if (!text) return [];
		if (limit <= 0) return this.baseChunker(text, limit);
		const chunkLimit = resolveQQBotMarkdownChunkLimit(limit);
		const chunks = [];
		const textWithPendingRow = this.consumePendingRowPrefix(text);
		const textWithPendingFenceLine = this.consumePendingFenceLinePrefix(textWithPendingRow);
		const hasTrailingNewline = textWithPendingFenceLine.endsWith("\n");
		const lines = textWithPendingFenceLine.split("\n");
		for (const [index, line] of lines.entries()) {
			const isTrailingSplitLine = index === lines.length - 1 && line === "";
			this.consumeLine(line, {
				limit: chunkLimit,
				chunks,
				hasTrailingNewline,
				isTrailingSplitLine,
				isLastLine: index === lines.length - 1
			});
		}
		this.flushText(chunks, chunkLimit);
		this.flushTable(chunks);
		return chunks;
	}
	flushPendingText(limit) {
		const chunkLimit = resolveQQBotMarkdownChunkLimit(limit);
		const chunks = [];
		this.flushPendingRowFragment(chunks, chunkLimit);
		this.flushPendingFenceLineFragment();
		this.flushPendingHeaderAsText();
		this.flushText(chunks, chunkLimit);
		this.flushTable(chunks);
		return chunks;
	}
	consumeLine(line, params) {
		const fence = parseFenceLine(line);
		if (fence) {
			this.endTable(params.chunks);
			if (!this.activeFence) {
				this.pushTextLine(line);
				this.activeFence = fence;
				this.clearPendingTableHeader();
				return;
			}
			if (isClosingFenceLine(line, this.activeFence)) {
				this.pushFenceTextLine(line);
				this.activeFence = null;
				this.clearPendingTableHeader();
				return;
			}
			this.pushFenceTextLine(line);
			this.clearPendingTableHeader();
			return;
		}
		if (this.activeFence) {
			if (params.isLastLine && !params.hasTrailingNewline) {
				this.pendingFenceLineFragment = mergeFenceLineFragments(this.pendingFenceLineFragment, line);
				return;
			}
			this.pushFenceTextLine(line);
			return;
		}
		if (isIncompleteTableRowFragment(line) || this.activeTable && isShortTableRowLine(line, this.activeTable)) {
			if (params.isLastLine) {
				this.flushText(params.chunks, params.limit);
				this.pendingRowFragment = mergeRowFragments(this.pendingRowFragment, line);
				return;
			}
			this.pushTextLine(renderMalformedPipeLineAsText(line));
			return;
		}
		if (this.pendingHeaderLine && isTableSeparatorLine(line)) {
			this.flushText(params.chunks, params.limit);
			this.activeTable = {
				header: this.pendingHeaderLine,
				separator: line,
				cells: this.pendingHeaderCells
			};
			this.pendingHeaderLine = null;
			this.pendingHeaderCells = [];
			this.ensureTableHeader();
			return;
		}
		if (isTableRowLine(line) && this.activeTable && !isTableSeparatorLine(line)) {
			this.flushText(params.chunks, params.limit);
			this.appendTableRow(line, params.limit, params.chunks);
			return;
		}
		if (this.activeTable) {
			if (!line.trim() && params.isTrailingSplitLine) return;
			this.endTable(params.chunks);
		}
		if (isTableRowLine(line) && !isTableSeparatorLine(line)) {
			this.flushText(params.chunks, params.limit);
			this.pendingHeaderLine = line;
			this.pendingHeaderCells = splitTableCells(line);
			return;
		}
		this.flushPendingHeaderAsText();
		this.pushTextLine(line);
	}
	pushTextLine(line) {
		this.textLines.push(line);
	}
	pushFenceTextLine(line) {
		if (this.textLines.length === 0 && this.activeFence) this.pendingTextFenceOpenLine = this.activeFence.openLine;
		this.textLines.push(line);
	}
	appendTableRow(line, limit, chunks) {
		if (utf8ByteLength([
			this.activeTable.header,
			this.activeTable.separator,
			line
		].join("\n")) > limit) {
			this.dropHeaderOnlyTableChunk();
			this.flushTable(chunks);
			this.pushOversizedTableRow(line, limit, chunks);
			return;
		}
		this.ensureTableHeader();
		if (utf8ByteLength([...this.tableLines, line].join("\n")) <= limit) {
			this.tableLines.push(line);
			return;
		}
		this.flushTable(chunks);
		this.ensureTableHeader();
		this.tableLines.push(line);
	}
	pushOversizedTableRow(line, limit, chunks) {
		pushBaseChunks(chunks, renderTableRowAsFields(this.activeTable.cells, splitTableCells(line)), limit, this.baseChunker);
	}
	ensureTableHeader() {
		if (this.tableLines.length > 0 || !this.activeTable) return;
		this.tableLines.push(this.activeTable.header, this.activeTable.separator);
	}
	flushText(chunks, limit) {
		if (this.textLines.length === 0) return;
		if (this.flushFenceText(chunks, limit)) return;
		let text = this.textLines.join("\n");
		this.textLines = [];
		if (this.pendingTextFenceOpenLine) {
			text = `${this.pendingTextFenceOpenLine}\n${text}`;
			this.pendingTextFenceOpenLine = null;
		}
		if (this.activeFence) text = `${text}\n${this.activeFence.closeLine}`;
		if (!text) return;
		pushBaseChunks(chunks, text, limit, this.baseChunker);
	}
	flushFenceText(chunks, limit) {
		const pendingFenceOpenLine = this.pendingTextFenceOpenLine;
		const firstLineFence = pendingFenceOpenLine ? null : parseFenceLine(this.textLines[0] ?? "");
		const fence = pendingFenceOpenLine ? parseFenceLine(pendingFenceOpenLine) : firstLineFence;
		if (!fence) return false;
		const bodyLines = pendingFenceOpenLine ? [...this.textLines] : this.textLines.slice(1);
		this.textLines = [];
		this.pendingTextFenceOpenLine = null;
		if (bodyLines.length > 0 && isClosingFenceLine(bodyLines[bodyLines.length - 1], fence)) bodyLines.pop();
		if (this.activeFence && bodyLines.length === 0) return true;
		pushFenceLineChunks({
			chunks,
			openLine: fence.openLine,
			closeLine: fence.closeLine,
			bodyLines,
			limit,
			baseChunker: this.baseChunker
		});
		return true;
	}
	consumePendingFenceLinePrefix(text) {
		if (!this.pendingFenceLineFragment) return text;
		const separator = shouldJoinFenceLineFragments(this.pendingFenceLineFragment, text) ? "" : "\n";
		const merged = `${this.pendingFenceLineFragment}${separator}${text}`;
		this.pendingFenceLineFragment = null;
		return merged;
	}
	flushPendingFenceLineFragment() {
		if (!this.pendingFenceLineFragment) return;
		this.pushFenceTextLine(this.pendingFenceLineFragment);
		this.pendingFenceLineFragment = null;
	}
	flushPendingHeaderAsText() {
		if (!this.pendingHeaderLine) return;
		this.pushTextLine(this.pendingHeaderLine);
		this.pendingHeaderLine = null;
		this.pendingHeaderCells = [];
	}
	clearPendingTableHeader() {
		this.pendingHeaderLine = null;
		this.pendingHeaderCells = [];
	}
	consumePendingRowPrefix(text) {
		if (!this.pendingRowFragment) return text;
		const separator = this.pendingRowFragment.trimEnd().endsWith("|") && text && !/^[\s|]/.test(text) ? " " : "";
		const merged = `${this.pendingRowFragment}${separator}${text}`;
		this.pendingRowFragment = null;
		return merged;
	}
	flushPendingRowFragment(chunks, limit) {
		if (!this.pendingRowFragment) return;
		const fragment = this.pendingRowFragment;
		this.pendingRowFragment = null;
		pushBaseChunks(chunks, this.activeTable ? renderTableRowAsFields(this.activeTable.cells, splitPartialTableCells(fragment)) : renderMalformedPipeLineAsText(fragment), limit, this.baseChunker);
	}
	flushTable(chunks) {
		if (this.tableLines.length === 0) return;
		chunks.push(this.tableLines.join("\n"));
		this.tableLines = [];
	}
	dropHeaderOnlyTableChunk() {
		if (this.activeTable && this.tableLines.length === 2 && this.tableLines[0] === this.activeTable.header && this.tableLines[1] === this.activeTable.separator) this.tableLines = [];
	}
	endTable(chunks) {
		this.flushTable(chunks);
		this.activeTable = null;
	}
};
function isTableRowLine(line) {
	const trimmed = line.trim();
	return trimmed.startsWith("|") && trimmed.endsWith("|") && splitTableCells(trimmed).length >= 2;
}
function resolveQQBotMarkdownChunkLimit(limit) {
	return Math.min(limit, QQBOT_MARKDOWN_SAFE_CHUNK_BYTE_LIMIT);
}
function pushBaseChunks(chunks, text, byteLimit, baseChunker) {
	for (const chunk of baseChunker(text, byteLimit)) {
		if (!chunk) continue;
		if (utf8ByteLength(chunk) <= byteLimit) {
			chunks.push(chunk);
			continue;
		}
		chunks.push(...splitByUtf8ByteLimit(chunk, byteLimit));
	}
}
function splitByUtf8ByteLimit(text, byteLimit) {
	if (!text) return [];
	const chunks = [];
	let current = "";
	let currentBytes = 0;
	for (const char of text) {
		const charBytes = utf8ByteLength(char);
		if (current && currentBytes + charBytes > byteLimit) {
			chunks.push(current);
			current = "";
			currentBytes = 0;
		}
		current += char;
		currentBytes += charBytes;
	}
	if (current) chunks.push(current);
	return chunks;
}
function utf8ByteLength(text) {
	return Buffer.byteLength(text, "utf8");
}
function isIncompleteTableRowFragment(line) {
	const trimmed = line.trim();
	return trimmed.startsWith("|") && !trimmed.endsWith("|") && splitPartialTableCells(trimmed).length >= 2;
}
function isShortTableRowLine(line, table) {
	if (!isTableRowLine(line) || isTableSeparatorLine(line)) return false;
	return splitTableCells(line).length < table.cells.length;
}
function isTableSeparatorLine(line) {
	if (!isTableRowLine(line)) return false;
	const cells = splitTableCells(line);
	return cells.length > 0 && cells.every((cell) => /^:?-+:?$/.test(cell.trim()));
}
function splitTableRowCells(inner) {
	const cells = [];
	let current = "";
	for (let i = 0; i < inner.length; i++) {
		const char = inner[i];
		if (char === "\\" && i + 1 < inner.length) {
			const next = inner[i + 1];
			current += next === "|" || next === "\\" ? next : `\\${next}`;
			i++;
			continue;
		}
		if (char === "|") {
			cells.push(current);
			current = "";
			continue;
		}
		current += char;
	}
	cells.push(current);
	return cells;
}
function splitTableCells(line) {
	return splitTableRowCells(line.trim().slice(1, -1)).map((cell) => cell.trim());
}
function splitPartialTableCells(line) {
	return splitTableRowCells(line.trim().replace(/^\|/, "")).map((cell) => cell.trim()).filter((cell) => cell.length > 0);
}
function mergeRowFragments(pending, next) {
	return pending ? `${pending}${next}` : next;
}
function mergeFenceLineFragments(pending, next) {
	return pending ? `${pending}${next}` : next;
}
function shouldJoinFenceLineFragments(pending, next) {
	if (!next || next.startsWith("\n")) return true;
	const trimmedPending = pending.trimEnd();
	const trimmedNext = next.trimStart();
	if (!trimmedPending || !trimmedNext) return true;
	if (/\d\.$/.test(trimmedPending) && /^\d/.test(trimmedNext)) return true;
	if (/[.([{:,+\-*/%=&|^<>\\]$/.test(trimmedPending)) return true;
	return hasUnclosedQuote(trimmedPending) || hasUnclosedDelimiter(trimmedPending);
}
function hasUnclosedQuote(line) {
	let single = false;
	let double = false;
	let escaped = false;
	for (const char of line) {
		if (escaped) {
			escaped = false;
			continue;
		}
		if (char === "\\") {
			escaped = true;
			continue;
		}
		if (char === "'" && !double) {
			single = !single;
			continue;
		}
		if (char === "\"" && !single) double = !double;
	}
	return single || double;
}
function hasUnclosedDelimiter(line) {
	const stack = [];
	const pairs = {
		"(": ")",
		"[": "]",
		"{": "}"
	};
	const closers = new Set(Object.values(pairs));
	for (const char of line) {
		if (pairs[char]) {
			stack.push(pairs[char]);
			continue;
		}
		if (closers.has(char)) {
			if (stack.at(-1) === char) stack.pop();
		}
	}
	return stack.length > 0;
}
function renderMalformedPipeLineAsText(line) {
	return splitPartialTableCells(line).join(" ");
}
function renderTableRowAsFields(headers, cells) {
	return cells.map((cell, index) => {
		const header = headers[index]?.trim();
		return header ? `${header}: ${cell}` : cell;
	}).join("\n");
}
function pushFenceLineChunks(params) {
	const { chunks, openLine, closeLine, bodyLines, limit, baseChunker } = params;
	let currentLines = [];
	const render = (lines) => [
		openLine,
		...lines,
		closeLine
	].join("\n");
	const flushCurrent = () => {
		if (currentLines.length === 0) return;
		chunks.push(render(currentLines));
		currentLines = [];
	};
	for (const line of bodyLines) {
		const candidate = [...currentLines, line];
		if (utf8ByteLength(render(candidate)) <= limit) {
			currentLines = candidate;
			continue;
		}
		flushCurrent();
		const singleLineChunk = render([line]);
		if (utf8ByteLength(singleLineChunk) <= limit) {
			currentLines = [line];
			continue;
		}
		pushBaseChunks(chunks, singleLineChunk, limit, baseChunker);
	}
	if (currentLines.length > 0 || bodyLines.length === 0) chunks.push(render(currentLines));
}
function parseFenceLine(line) {
	const match = line.match(/^(\s*)(`{3,}|~{3,})/);
	if (!match?.[2]) return null;
	return {
		openLine: line,
		closeLine: `${match[1] ?? ""}${match[2]}`,
		marker: match[2]
	};
}
function isClosingFenceLine(line, fence) {
	const markerChar = fence.marker[0] === "`" ? "`" : "~";
	const match = line.match(/^(\s*)(`{3,}|~{3,})\s*$/);
	return Boolean(match?.[2] && match[2][0] === markerChar && match[2].length >= fence.marker.length);
}
//#endregion
//#region extensions/qqbot/src/group-policy.ts
function resolveQQBotGroupToolPolicy(params) {
	return resolveChannelGroupToolsPolicy({
		cfg: params.cfg,
		channel: "qqbot",
		groupId: params.groupId,
		groupIdCaseInsensitive: true,
		accountId: params.accountId,
		senderId: params.senderId,
		senderName: params.senderName,
		senderUsername: params.senderUsername,
		senderE164: params.senderE164
	});
}
//#endregion
//#region extensions/qqbot/src/channel.ts
const loadGatewayModule = createLazyRuntimeModule(() => import("./gateway-pJQppxe4.js"));
const loadOutboundMessagingModule = createLazyRuntimeModule(() => import("./outbound-BIrfvvFJ.js").then((n) => n.t));
function createQQBotSendReceipt(params) {
	const messageId = params.messageId?.trim();
	return createMessageReceiptFromOutboundResults({
		results: messageId ? [{
			channel: "qqbot",
			messageId,
			conversationId: params.target
		}] : [],
		threadId: params.target,
		kind: params.kind
	});
}
function resolveQQBotOutboundSessionRoute(params) {
	const target = parseTarget(params.target);
	const chatType = target.type === "c2c" ? "direct" : "group";
	const qualifiedTarget = `qqbot:${target.type}:${target.id}`;
	return buildChannelOutboundSessionRoute({
		cfg: params.cfg,
		agentId: params.agentId,
		channel: "qqbot",
		accountId: params.accountId,
		recipientSessionExact: true,
		peer: {
			kind: chatType,
			id: target.id
		},
		chatType,
		from: qualifiedTarget,
		to: qualifiedTarget
	});
}
async function sendQQBotText(params) {
	await loadGatewayModule();
	const account = resolveQQBotAccount(params.cfg, params.accountId);
	const { sendText } = await loadOutboundMessagingModule();
	const result = await sendText({
		to: params.to,
		text: params.text,
		accountId: params.accountId,
		replyToId: params.replyToId,
		account: toGatewayAccount(account),
		...params.mediaAccess ? { mediaAccess: params.mediaAccess } : {},
		...params.mediaLocalRoots ? { mediaLocalRoots: params.mediaLocalRoots } : {},
		...params.mediaReadFile ? { mediaReadFile: params.mediaReadFile } : {}
	});
	return {
		channel: "qqbot",
		messageId: result.messageId ?? "",
		receipt: createQQBotSendReceipt({
			messageId: result.messageId,
			target: params.to,
			kind: "text"
		}),
		meta: result.error ? { error: result.error } : void 0
	};
}
async function sendQQBotMedia(params) {
	await loadGatewayModule();
	const account = resolveQQBotAccount(params.cfg, params.accountId);
	const { sendMedia } = await loadOutboundMessagingModule();
	const result = await sendMedia({
		to: params.to,
		text: params.text ?? "",
		mediaUrl: params.mediaUrl ?? "",
		accountId: params.accountId,
		replyToId: params.replyToId,
		account: toGatewayAccount(account),
		...params.mediaAccess ? { mediaAccess: params.mediaAccess } : {},
		...params.mediaLocalRoots ? { mediaLocalRoots: params.mediaLocalRoots } : {},
		...params.mediaReadFile ? { mediaReadFile: params.mediaReadFile } : {}
	});
	return {
		channel: "qqbot",
		messageId: result.messageId ?? "",
		receipt: createQQBotSendReceipt({
			messageId: result.messageId,
			target: params.to,
			kind: "media"
		}),
		meta: result.error ? { error: result.error } : void 0
	};
}
function resolveQQBotOutboundMediaAccessContext(ctx) {
	const record = ctx && typeof ctx === "object" ? ctx : void 0;
	return {
		...record?.mediaAccess ? { mediaAccess: record.mediaAccess } : {},
		...record?.mediaLocalRoots ? { mediaLocalRoots: record.mediaLocalRoots } : {},
		...record?.mediaReadFile ? { mediaReadFile: record.mediaReadFile } : {}
	};
}
function toQQBotMessageSendResult(result) {
	if (result.meta?.error) throw new Error(result.meta.error);
	if (result.receipt.platformMessageIds.length === 0) throw new Error("QQBot message adapter send did not return a platform message id");
	return {
		messageId: result.messageId || result.receipt.primaryPlatformMessageId,
		receipt: result.receipt
	};
}
const qqbotMessageAdapter = defineChannelMessageAdapter({
	id: "qqbot",
	durableFinal: { capabilities: {
		text: true,
		media: true,
		replyTo: true
	} },
	send: {
		text: async (ctx) => toQQBotMessageSendResult(await sendQQBotText({
			cfg: ctx.cfg,
			to: ctx.to,
			text: ctx.text,
			accountId: ctx.accountId,
			replyToId: ctx.replyToId,
			...resolveQQBotOutboundMediaAccessContext(ctx)
		})),
		media: async (ctx) => toQQBotMessageSendResult(await sendQQBotMedia({
			cfg: ctx.cfg,
			to: ctx.to,
			text: ctx.text,
			mediaUrl: ctx.mediaUrl,
			accountId: ctx.accountId,
			replyToId: ctx.replyToId,
			...resolveQQBotOutboundMediaAccessContext(ctx)
		}))
	}
});
const EXEC_APPROVAL_COMMAND_RE = /\/approve(?:@[^\s]+)?\s+[A-Za-z0-9][A-Za-z0-9._:-]*\s+(?:allow-once|allow-always|always|deny)\b/i;
function persistAccountCredentialSnapshot(account) {
	if (account.appId && account.clientSecret) saveCredentialBackup(account.accountId, account.appId, account.clientSecret);
}
function shouldSuppressLocalQQBotApprovalPrompt(params) {
	if (params.hint?.kind !== "approval-pending" || params.hint.approvalKind !== "exec") return false;
	const account = resolveQQBotAccount(params.cfg, params.accountId);
	if (!account.enabled || account.secretSource === "none") return false;
	if (getExecApprovalReplyMetadata(params.payload)) return true;
	const text = typeof params.payload.text === "string" ? params.payload.text : "";
	return EXEC_APPROVAL_COMMAND_RE.test(text);
}
const qqbotPlugin = {
	id: "qqbot",
	setupWizard: qqbotSetupWizard,
	meta: { ...qqbotMeta },
	capabilities: {
		chatTypes: ["direct", "group"],
		media: true,
		reactions: false,
		threads: false,
		blockStreaming: true
	},
	reload: { configPrefixes: ["channels.qqbot"] },
	configSchema: qqbotChannelConfigSchema,
	doctor: qqbotDoctor,
	config: {
		...qqbotConfigAdapter,
		/**
		* Treat an account as configured when either the live config has
		* credentials OR a recoverable credential backup exists. This mirrors
		* the standalone plugin and lets the gateway survive a hot upgrade
		* that wiped openclaw.json mid-flight.
		*/
		isConfigured: (account) => {
			if (qqbotConfigAdapter.isConfigured(account)) return true;
			if (!account) return false;
			const backup = loadCredentialBackup(account.accountId);
			return Boolean(backup?.appId && backup?.clientSecret);
		}
	},
	setup: { ...qqbotSetupAdapterShared },
	approvalCapability: getQQBotApprovalCapability(),
	groups: { resolveToolPolicy: resolveQQBotGroupToolPolicy },
	message: qqbotMessageAdapter,
	messaging: {
		targetPrefixes: ["qqbot"],
		/** Normalize common QQ Bot target formats into the canonical qqbot:... form. */
		normalizeTarget,
		resolveOutboundSessionRoute: (params) => resolveQQBotOutboundSessionRoute(params),
		targetResolver: {
			/** Return true when the id looks like a QQ Bot target. */
			looksLikeId: looksLikeQQBotTarget,
			hint: "QQ Bot target format: qqbot:c2c:openid (direct) or qqbot:group:groupid (group)"
		}
	},
	outbound: {
		deliveryMode: "direct",
		chunker: (text, limit) => chunkQQBotMarkdownText(text, limit, getQQBotRuntime().channel.text.chunkMarkdownText),
		chunkerMode: "markdown",
		textChunkLimit: 5e3,
		sanitizeText: ({ text }) => sanitizeAssistantVisibleText(text),
		shouldSuppressLocalPayloadPrompt: ({ cfg, accountId, payload, hint }) => shouldSuppressLocalQQBotApprovalPrompt({
			cfg,
			accountId,
			payload,
			hint
		}),
		sendText: async (ctx) => await sendQQBotText({
			cfg: ctx.cfg,
			to: ctx.to,
			text: ctx.text,
			accountId: ctx.accountId,
			replyToId: ctx.replyToId,
			...resolveQQBotOutboundMediaAccessContext(ctx)
		}),
		sendMedia: async (ctx) => await sendQQBotMedia({
			cfg: ctx.cfg,
			to: ctx.to,
			text: ctx.text,
			mediaUrl: ctx.mediaUrl,
			accountId: ctx.accountId,
			replyToId: ctx.replyToId,
			...resolveQQBotOutboundMediaAccessContext(ctx)
		})
	},
	gateway: {
		startAccount: async (ctx) => {
			let { account, cfg } = ctx;
			const { abortSignal, log } = ctx;
			if (!account.appId || !account.clientSecret) {
				const backup = loadCredentialBackup(account.accountId);
				if (backup?.appId && backup?.clientSecret) try {
					const nextCfg = applyQQBotAccountConfig(cfg, account.accountId, {
						appId: backup.appId,
						clientSecret: backup.clientSecret
					});
					await writeOpenClawConfigThroughRuntime(getQQBotRuntime(), nextCfg);
					cfg = nextCfg;
					account = resolveQQBotAccount(nextCfg, account.accountId);
					log?.info(`[qqbot:${account.accountId}] Restored credentials from backup (appId=${account.appId})`);
				} catch (err) {
					log?.error(`[qqbot:${account.accountId}] Failed to restore credentials from backup: ${err instanceof Error ? err.message : String(err)}`);
				}
			}
			const { startGateway } = await loadGatewayModule();
			log?.info(`[qqbot:${account.accountId}] Starting gateway — appId=${account.appId}, enabled=${account.enabled}, name=${account.name ?? "unnamed"}`);
			await startGateway({
				account,
				abortSignal,
				cfg,
				log,
				channelRuntime: ctx.channelRuntime,
				onReady: () => {
					log?.info(`[qqbot:${account.accountId}] Gateway ready`);
					ctx.setStatus({
						...ctx.getStatus(),
						running: true,
						connected: true,
						lastConnectedAt: Date.now(),
						lastError: null
					});
					persistAccountCredentialSnapshot(account);
				},
				onResumed: () => {
					log?.info(`[qqbot:${account.accountId}] Gateway resumed`);
					ctx.setStatus({
						...ctx.getStatus(),
						running: true,
						connected: true,
						lastConnectedAt: Date.now(),
						lastError: null
					});
					persistAccountCredentialSnapshot(account);
				},
				onError: (error) => {
					log?.error(`[qqbot:${account.accountId}] Gateway error: ${error.message}`);
					ctx.setStatus({
						...ctx.getStatus(),
						lastError: error.message
					});
				},
				onDisconnected: ({ reason, fatal }) => {
					log?.info(`[qqbot:${account.accountId}] Gateway disconnected${reason ? `: ${reason}` : ""}`);
					ctx.setStatus({
						...ctx.getStatus(),
						connected: false,
						...fatal && reason ? { lastError: reason } : {}
					});
				}
			});
		},
		logoutAccount: async ({ accountId, cfg }) => {
			const { nextCfg, cleared, changed } = clearAccountCredentials(cfg, accountId);
			if (changed) await writeOpenClawConfigThroughRuntime(getQQBotRuntime(), nextCfg);
			const loggedOut = resolveQQBotAccount(changed ? nextCfg : cfg, accountId).secretSource === "none";
			return {
				ok: true,
				cleared,
				envToken: Boolean(process.env.QQBOT_CLIENT_SECRET),
				loggedOut
			};
		}
	},
	status: {
		defaultRuntime: {
			accountId: DEFAULT_ACCOUNT_ID$1,
			running: false,
			connected: false,
			lastConnectedAt: null,
			lastError: null,
			lastInboundAt: null,
			lastOutboundAt: null
		},
		buildChannelSummary: ({ snapshot }) => ({
			configured: snapshot.configured ?? false,
			tokenSource: snapshot.tokenSource ?? "none",
			running: snapshot.running ?? false,
			connected: snapshot.connected ?? false,
			lastConnectedAt: snapshot.lastConnectedAt ?? null,
			lastError: snapshot.lastError ?? null
		}),
		buildAccountSnapshot: ({ account, runtime }) => ({
			accountId: account?.accountId ?? DEFAULT_ACCOUNT_ID$1,
			name: account?.name,
			enabled: account?.enabled ?? false,
			configured: Boolean(account?.appId && account?.clientSecret),
			tokenSource: account?.secretSource,
			running: runtime?.running ?? false,
			connected: runtime?.connected ?? false,
			lastConnectedAt: runtime?.lastConnectedAt ?? null,
			lastError: runtime?.lastError ?? null,
			lastInboundAt: runtime?.lastInboundAt ?? null,
			lastOutboundAt: runtime?.lastOutboundAt ?? null
		})
	}
};
//#endregion
export { authorizeQQBotApprovalAction as a, resolveQQBotExecApprovalConfig as c, buildExecApprovalText as d, buildPluginApprovalText as f, toGatewayAccount as i, shouldHandleQQBotExecApprovalRequest as l, resolveApprovalTarget as m, createQQBotMarkdownChunker as n, isQQBotExecApprovalClientEnabled as o, parseApprovalButtonData as p, openQQBotSyncKeyedStore as r, matchesQQBotApprovalAccount as s, qqbotPlugin as t, buildApprovalKeyboard as u };
