// @openclaw/qqbot/dist/channel-entry-C5YdhX3Y.js
import { a as resolveQQBotAccount } from "./config-C1qZbh0K.js";
import { d as PRIVATE_CHAT_ONLY_TEXT, n as resolveGroupCommandLevelFromAccountConfig, o as getFrameworkCommands } from "./group-o0GmovSf.js";
import { h as sendDocument } from "./outbound-BIrfvvFJ.js";
import { t as registerQQBotTools } from "./tools-UJJ-tLHP.js";
//#region extensions/qqbot/src/bridge/commands/framework-context-adapter.ts
/**
* Default queue snapshot used for framework-registered commands.
*
* Framework-side command dispatch runs outside the per-sender queue, so
* handlers observe an empty snapshot by design.
*/
const DEFAULT_QUEUE_SNAPSHOT = {
	totalPending: 0,
	activeUsers: 0,
	maxConcurrentUsers: 10,
	senderPending: 0
};
function buildFrameworkSlashContext({ ctx, account, from, commandName, groupCommandLevel }) {
	const args = ctx.args ?? "";
	const rawContent = args ? `/${commandName} ${args}` : `/${commandName}`;
	return {
		type: from.msgType,
		senderId: ctx.senderId ?? "",
		messageId: "",
		eventTimestamp: (/* @__PURE__ */ new Date()).toISOString(),
		receivedAt: Date.now(),
		rawContent,
		args,
		accountId: account.accountId,
		appId: account.appId,
		accountConfig: account.config,
		commandAuthorized: ctx.isAuthorizedSender,
		groupCommandLevel,
		queueSnapshot: { ...DEFAULT_QUEUE_SNAPSHOT }
	};
}
//#endregion
//#region extensions/qqbot/src/bridge/commands/from-parser.ts
const MSG_TYPE_MAP = {
	c2c: "c2c",
	dm: "dm",
	group: "group",
	channel: "guild"
};
const TARGET_TYPE_MAP = {
	c2c: "c2c",
	dm: "dm",
	group: "group",
	channel: "channel"
};
function isFromKind(value) {
	return value === "c2c" || value === "dm" || value === "group" || value === "channel";
}
/**
* Parse `ctx.from` into the structured fields the QQBot bridge expects.
*
* Unknown or missing prefixes fall back to c2c. The remainder after the first
* `:` is returned verbatim as the target id, matching what the previous inline
* implementation did.
*/
function parseQQBotFrom(from) {
	const stripped = (from ?? "").replace(/^qqbot:/iu, "");
	const colonIdx = stripped.indexOf(":");
	const rawPrefix = colonIdx === -1 ? stripped : stripped.slice(0, colonIdx);
	const targetId = colonIdx === -1 ? stripped : stripped.slice(colonIdx + 1);
	const kind = isFromKind(rawPrefix) ? rawPrefix : "c2c";
	return {
		msgType: MSG_TYPE_MAP[kind],
		targetType: TARGET_TYPE_MAP[kind],
		targetId
	};
}
//#endregion
//#region extensions/qqbot/src/bridge/commands/result-dispatcher.ts
const UNEXPECTED_RESULT_TEXT = "⚠️ 命令返回了意外结果。";
function hasFilePath(value) {
	return typeof value === "object" && value !== null && "filePath" in value && typeof value.filePath === "string";
}
function buildMediaTarget(account, from) {
	return {
		targetType: from.targetType,
		targetId: from.targetId,
		account
	};
}
async function dispatchFrameworkSlashResult({ result, account, from, logger }) {
	if (typeof result === "string") return { text: result };
	if (hasFilePath(result)) {
		const mediaCtx = buildMediaTarget(account, from);
		try {
			await sendDocument(mediaCtx, result.filePath, { allowQQBotDataDownloads: true });
		} catch (err) {
			logger?.warn(`framework slash file send failed: ${String(err)}`);
		}
		return { text: result.text };
	}
	return { text: UNEXPECTED_RESULT_TEXT };
}
//#endregion
//#region extensions/qqbot/src/bridge/commands/framework-registration.ts
function isExplicitQQBotC2cFrom(from) {
	const raw = (from ?? "").trim();
	const stripped = raw.replace(/^qqbot:/iu, "");
	const colonIdx = stripped.indexOf(":");
	if (colonIdx === -1) return false;
	const kind = stripped.slice(0, colonIdx).toLowerCase();
	const targetId = stripped.slice(colonIdx + 1).trim();
	return /^qqbot:/iu.test(raw) && kind === "c2c" && targetId.length > 0;
}
function registerQQBotFrameworkCommands(api) {
	for (const cmd of getFrameworkCommands()) api.registerCommand({
		name: cmd.name,
		description: cmd.description,
		channels: ["qqbot"],
		requireAuth: true,
		acceptsArgs: true,
		handler: async (ctx) => {
			const from = parseQQBotFrom(ctx.from);
			const account = resolveQQBotAccount(ctx.config, ctx.accountId ?? void 0);
			const groupCommandLevel = from.msgType === "group" || from.msgType === "guild" ? resolveGroupCommandLevelFromAccountConfig(account.config, from.targetId) : void 0;
			if (cmd.c2cOnly && !isExplicitQQBotC2cFrom(ctx.from)) return { text: PRIVATE_CHAT_ONLY_TEXT };
			const slashCtx = buildFrameworkSlashContext({
				ctx,
				account,
				from,
				commandName: cmd.name,
				groupCommandLevel
			});
			return await dispatchFrameworkSlashResult({
				result: await cmd.handler(slashCtx),
				account,
				from,
				logger: api.logger
			});
		}
	});
}
//#endregion
//#region extensions/qqbot/src/bridge/channel-entry.ts
function registerQQBotFull(api) {
	registerQQBotTools(api);
	registerQQBotFrameworkCommands(api);
}
//#endregion
export { registerQQBotFull as t };
