// @openclaw/qqbot/dist/doctor-contract-api.js
import { n as legacyConfigRules, r as normalizeCompatibilityConfig, t as buildQQBotStateKey } from "./state-keys-CQKlAFyo.js";
import path from "node:path";
import fs from "node:fs/promises";
//#region extensions/qqbot/src/state-migrations.ts
const CREDENTIAL_BACKUPS_NAMESPACE = "credential-backups";
const MAX_CREDENTIAL_BACKUPS = 1e3;
function safeName(id) {
	return id.replace(/[^a-zA-Z0-9._-]/g, "_");
}
async function fileExists(filePath) {
	try {
		return (await fs.lstat(filePath)).isFile();
	} catch {
		return false;
	}
}
async function readCredentialBackup(filePath) {
	try {
		const parsed = JSON.parse(await fs.readFile(filePath, "utf8"));
		if (typeof parsed.accountId !== "string" || typeof parsed.appId !== "string" || typeof parsed.clientSecret !== "string" || !parsed.accountId || !parsed.appId || !parsed.clientSecret) return null;
		return {
			accountId: parsed.accountId,
			appId: parsed.appId,
			clientSecret: parsed.clientSecret,
			savedAt: typeof parsed.savedAt === "string" && parsed.savedAt ? parsed.savedAt : (/* @__PURE__ */ new Date(0)).toISOString()
		};
	} catch {
		return null;
	}
}
function credentialBackupKey(accountId) {
	return buildQQBotStateKey("credential-backup", accountId);
}
async function credentialBackupCandidates(stateDir) {
	const dataDir = path.join(stateDir, "qqbot", "data");
	const accountFiles = [];
	try {
		for (const entry of await fs.readdir(dataDir, { withFileTypes: true })) if (entry.isFile() && entry.name.startsWith("credential-backup-") && entry.name.endsWith(".json")) accountFiles.push({
			sourcePath: path.join(dataDir, entry.name),
			expectedSafeAccountId: entry.name.slice(18, -5)
		});
	} catch {}
	accountFiles.sort((left, right) => left.sourcePath.localeCompare(right.sourcePath));
	const singlePath = path.join(dataDir, "credential-backup.json");
	return await fileExists(singlePath) ? [...accountFiles, { sourcePath: singlePath }] : accountFiles;
}
async function readLegacyCredentialBackups(stateDir) {
	const backups = [];
	for (const candidate of await credentialBackupCandidates(stateDir)) {
		const value = await readCredentialBackup(candidate.sourcePath);
		if (!value || candidate.expectedSafeAccountId !== void 0 && safeName(value.accountId) !== candidate.expectedSafeAccountId) continue;
		backups.push({
			sourcePath: candidate.sourcePath,
			key: credentialBackupKey(value.accountId),
			value
		});
	}
	return backups;
}
async function archiveLegacySource(params) {
	const archivedPath = `${params.sourcePath}.migrated`;
	if (await fileExists(archivedPath)) {
		params.warnings.push(`Left QQBot credential backup in place because ${archivedPath} already exists`);
		return;
	}
	try {
		await fs.chmod(params.sourcePath, 384);
	} catch (err) {
		params.warnings.push(`Failed securing QQBot credential backup legacy source: ${String(err)}`);
		return;
	}
	try {
		await fs.rename(params.sourcePath, archivedPath);
		try {
			await fs.chmod(archivedPath, 384);
		} catch (err) {
			params.warnings.push(`Failed securing archived QQBot credential backup legacy source: ${String(err)}`);
		}
		params.changes.push(`Archived QQBot credential backup legacy source -> ${archivedPath}`);
	} catch (err) {
		params.warnings.push(`Failed archiving QQBot credential backup: ${String(err)}`);
	}
}
function sameCredentialBackup(left, right) {
	return left?.accountId === right.accountId && left.appId === right.appId && left.clientSecret === right.clientSecret && left.savedAt === right.savedAt;
}
async function rollbackCredentialImports(store, inserted, existing) {
	for (const [key, value] of [...inserted].toReversed()) if (sameCredentialBackup(await store.lookup(key), value)) await store.delete(key);
	for (const [key, value] of existing) if (await store.lookup(key) === void 0) await store.registerIfAbsent(key, value);
}
function findMissingKey(expected, actual) {
	for (const key of expected) if (!actual.has(key)) return key;
	return null;
}
const stateMigrations = [{
	id: "qqbot-credential-backups-json-to-plugin-state",
	label: "QQBot credential backups",
	async detectLegacyState(params) {
		const backups = await readLegacyCredentialBackups(params.stateDir);
		if (backups.length === 0) return null;
		return { preview: [`- QQBot credential backups: ${backups.length} ${backups.length === 1 ? "file" : "files"} -> plugin state (${CREDENTIAL_BACKUPS_NAMESPACE})`] };
	},
	async migrateLegacyState(params) {
		const changes = [];
		const warnings = [];
		const backups = await readLegacyCredentialBackups(params.stateDir);
		if (backups.length === 0) return {
			changes,
			warnings
		};
		const selectedByKey = /* @__PURE__ */ new Map();
		for (const backup of backups) if (!selectedByKey.has(backup.key)) selectedByKey.set(backup.key, backup);
		const store = params.context.openPluginStateKeyedStore({
			namespace: CREDENTIAL_BACKUPS_NAMESPACE,
			maxEntries: MAX_CREDENTIAL_BACKUPS
		});
		const existingEntries = await store.entries();
		const existingValues = new Map(existingEntries.map((entry) => [entry.key, entry.value]));
		const existingKeys = new Set(existingValues.keys());
		const missing = [...selectedByKey.values()].filter((backup) => !existingKeys.has(backup.key));
		const available = MAX_CREDENTIAL_BACKUPS - existingKeys.size;
		if (missing.length > available) {
			warnings.push(`Skipped QQBot credential backup migration because plugin state has room for ${available} of ${missing.length} missing entries; left legacy sources in place`);
			return {
				changes,
				warnings
			};
		}
		const expectedKeys = new Set(existingKeys);
		const inserted = /* @__PURE__ */ new Map();
		for (const backup of missing) try {
			if (await store.registerIfAbsent(backup.key, backup.value)) inserted.set(backup.key, backup.value);
			const missingKey = findMissingKey(new Set(expectedKeys).add(backup.key), new Set((await store.entries()).map((entry) => entry.key)));
			if (missingKey) {
				await rollbackCredentialImports(store, inserted, existingValues);
				warnings.push(`Stopped QQBot credential backup migration because plugin state capacity evicted ${missingKey}; restored credential state and left legacy sources in place`);
				return {
					changes,
					warnings
				};
			}
			expectedKeys.add(backup.key);
		} catch (err) {
			try {
				await rollbackCredentialImports(store, inserted, existingValues);
			} catch (rollbackErr) {
				warnings.push(`Failed restoring QQBot credential state after migration error: ${String(rollbackErr)}`);
			}
			warnings.push(`Failed migrating QQBot credential backup: ${String(err)}; left legacy sources in place`);
			return {
				changes,
				warnings
			};
		}
		if (inserted.size > 0) changes.push(`Migrated ${inserted.size} QQBot credential ${inserted.size === 1 ? "backup" : "backups"} -> plugin state`);
		for (const backup of backups) await archiveLegacySource({
			sourcePath: backup.sourcePath,
			changes,
			warnings
		});
		return {
			changes,
			warnings
		};
	}
}];
//#endregion
export { legacyConfigRules, normalizeCompatibilityConfig, stateMigrations };
