// @openclaw/qqbot/dist/group-o0GmovSf.js
import { t as asOptionalObjectRecord } from "./string-normalize-R_0cKO7Q.js";
import { d as resolveAccountBase } from "./config-C1qZbh0K.js";
import { n as debugLog } from "./log-DEtcoDWe.js";
import { I as getQQBotMediaPath, N as getHomeDir, P as getQQBotDataDir, z as isWindows } from "./outbound-BIrfvvFJ.js";
import { asBoolean, uniqueStrings } from "openclaw/plugin-sdk/string-coerce-runtime";
import fs from "node:fs";
import path from "node:path";
import { formatByteSize } from "openclaw/plugin-sdk/number-runtime";
import { clearRuntimeConfigSnapshot } from "openclaw/plugin-sdk/runtime-config-snapshot";
import { loadJsonFile } from "openclaw/plugin-sdk/json-store";
//#region extensions/qqbot/src/engine/commands/builtin/state.ts
let resolveVersionGetter = () => "unknown";
let approveRuntimeGetter = null;
let PLUGIN_VERSION = "unknown";
/**
* Initialize command dependencies from the EngineAdapters.commands port.
* Called once by the bridge layer during startup.
*/
function initSlashCommandDeps(port) {
	resolveVersionGetter = port.resolveVersion;
	PLUGIN_VERSION = port.pluginVersion;
	approveRuntimeGetter = port.approveRuntimeGetter ?? null;
}
function resolveRuntimeServiceVersion() {
	return resolveVersionGetter();
}
function getPluginVersionString() {
	return PLUGIN_VERSION;
}
function getFrameworkVersionString() {
	return resolveVersionGetter();
}
function getApproveRuntimeGetter() {
	return approveRuntimeGetter;
}
//#endregion
//#region extensions/qqbot/src/engine/commands/builtin/register-approve.ts
function registerApproveCommands(registry) {
	registry.register({
		name: "bot-approve",
		description: "管理命令执行审批配置",
		requireAuth: true,
		c2cOnly: true,
		usage: [
			`/bot-approve            查看操作指引`,
			`/bot-approve on         开启审批（白名单模式，推荐）`,
			`/bot-approve off        关闭审批，命令直接执行`,
			`/bot-approve always     始终审批，每次执行都需审批`,
			`/bot-approve reset      恢复框架默认值`,
			`/bot-approve status     查看当前审批配置`
		].join("\n"),
		handler: async (ctx) => {
			const arg = ctx.args.trim().toLowerCase();
			let runtime;
			try {
				const getter = getApproveRuntimeGetter();
				if (!getter) throw new Error("runtime not available");
				runtime = getter();
			} catch {
				return [
					`🔐 命令执行审批配置`,
					``,
					`❌ 当前环境不支持在线配置修改，请通过 CLI 手动配置：`,
					``,
					`\`\`\`shell`,
					`# 开启审批（白名单模式）`,
					`openclaw config set tools.exec.security allowlist`,
					`openclaw config set tools.exec.ask on-miss`,
					``,
					`# 关闭审批`,
					`openclaw config set tools.exec.security full`,
					`openclaw config set tools.exec.ask off`,
					`\`\`\``
				].join("\n");
			}
			const configApi = runtime.config;
			const loadExecConfig = () => {
				const exec = (configApi.current().tools ?? {}).exec ?? {};
				return {
					security: typeof exec.security === "string" ? exec.security : "deny",
					ask: typeof exec.ask === "string" ? exec.ask : "on-miss"
				};
			};
			const writeExecConfig = async (security, ask) => {
				const cfg = structuredClone(configApi.current());
				const tools = cfg.tools ?? {};
				const exec = tools.exec ?? {};
				exec.security = security;
				exec.ask = ask;
				tools.exec = exec;
				cfg.tools = tools;
				await configApi.replaceConfigFile({
					nextConfig: cfg,
					afterWrite: { mode: "auto" }
				});
			};
			const formatStatus = (security, ask) => {
				const secIcon = security === "full" ? "🟢" : security === "allowlist" ? "🟡" : "🔴";
				const askIcon = ask === "off" ? "🟢" : ask === "always" ? "🔴" : "🟡";
				return [
					`🔐 当前审批配置`,
					``,
					`${secIcon} 安全模式 (security): **${security}**`,
					`${askIcon} 审批模式 (ask): **${ask}**`,
					``,
					security === "deny" ? `⚠️ 当前为 deny 模式，所有命令执行被拒绝` : security === "full" && ask === "off" ? `✅ 所有命令无需审批直接执行` : security === "allowlist" && ask === "on-miss" ? `🛡️ 白名单命令直接执行，其余需审批` : ask === "always" ? `🔒 每次命令执行都需要人工审批` : `ℹ️ security=${security}, ask=${ask}`
				].join("\n");
			};
			if (!arg) return [
				`🔐 命令执行审批配置`,
				``,
				`<qqbot-cmd-input text="/bot-approve on" show="/bot-approve on"/> 开启审批（白名单模式）`,
				`<qqbot-cmd-input text="/bot-approve off" show="/bot-approve off"/> 关闭审批`,
				`<qqbot-cmd-input text="/bot-approve always" show="/bot-approve always"/> 严格模式`,
				`<qqbot-cmd-input text="/bot-approve reset" show="/bot-approve reset"/> 恢复默认`,
				`<qqbot-cmd-input text="/bot-approve status" show="/bot-approve status"/> 查看当前配置`
			].join("\n");
			if (arg === "status") {
				const { security, ask } = loadExecConfig();
				return [
					formatStatus(security, ask),
					``,
					`<qqbot-cmd-input text="/bot-approve on" show="/bot-approve on"/> 开启审批`,
					`<qqbot-cmd-input text="/bot-approve off" show="/bot-approve off"/> 关闭审批`,
					`<qqbot-cmd-input text="/bot-approve always" show="/bot-approve always"/> 严格模式`,
					`<qqbot-cmd-input text="/bot-approve reset" show="/bot-approve reset"/> 恢复默认`
				].join("\n");
			}
			if (arg === "on") try {
				await writeExecConfig("allowlist", "on-miss");
				return [
					`✅ 审批已开启`,
					``,
					`• security = allowlist（白名单模式）`,
					`• ask = on-miss（未命中白名单时需审批）`,
					``,
					`已批准的命令自动加入白名单，下次直接执行。`
				].join("\n");
			} catch (err) {
				return `❌ 配置更新失败: ${err instanceof Error ? err.message : String(err)}`;
			}
			if (arg === "off") try {
				await writeExecConfig("full", "off");
				return [
					`✅ 审批已关闭`,
					``,
					`• security = full（允许所有命令）`,
					`• ask = off（不需要审批）`,
					``,
					`⚠️ 所有命令将直接执行，不会弹出审批确认。`
				].join("\n");
			} catch (err) {
				return `❌ 配置更新失败: ${err instanceof Error ? err.message : String(err)}`;
			}
			if (arg === "always" || arg === "strict") try {
				await writeExecConfig("allowlist", "always");
				return [
					`✅ 已切换为严格审批模式`,
					``,
					`• security = allowlist`,
					`• ask = always（每次执行都需审批）`,
					``,
					`每个命令都会弹出审批按钮，需手动确认。`
				].join("\n");
			} catch (err) {
				return `❌ 配置更新失败: ${err instanceof Error ? err.message : String(err)}`;
			}
			if (arg === "reset") try {
				const cfg = structuredClone(configApi.current());
				const tools = cfg.tools ?? {};
				const exec = tools.exec ?? {};
				delete exec.security;
				delete exec.ask;
				if (Object.keys(exec).length === 0) delete tools.exec;
				else tools.exec = exec;
				if (Object.keys(tools).length === 0) delete cfg.tools;
				else cfg.tools = tools;
				await configApi.replaceConfigFile({
					nextConfig: cfg,
					afterWrite: { mode: "auto" }
				});
				return [
					`✅ 审批配置已重置`,
					``,
					`已移除 tools.exec.security 和 tools.exec.ask`,
					`框架将使用默认值（security=deny, ask=on-miss）`,
					``,
					`如需开启命令执行，请使用 /bot-approve on`
				].join("\n");
			} catch (err) {
				return `❌ 配置更新失败: ${err instanceof Error ? err.message : String(err)}`;
			}
			return [
				`❌ 未知参数: ${arg}`,
				``,
				`可用选项: on | off | always | reset | status`,
				`输入 /bot-approve ? 查看详细用法`
			].join("\n");
		}
	});
}
//#endregion
//#region extensions/qqbot/src/engine/commands/builtin/register-basic.ts
const QQBOT_PLUGIN_GITHUB_URL = "https://github.com/openclaw/openclaw/tree/main/extensions/qqbot";
const QQBOT_UPGRADE_GUIDE_URL = "https://q.qq.com/qqbot/openclaw/upgrade.html";
function registerBasicBotCommands(registry) {
	registry.register({
		name: "bot-help",
		description: "查看所有内置命令",
		usage: [
			`/bot-help`,
			``,
			`查看所有可用的 QQBot 内置命令及其简要说明。`,
			`在命令后追加 ? 可查看详细用法。`
		].join("\n"),
		handler: (ctx) => {
			const isGroup = ctx.type === "group";
			const lines = [`### QQBot 内置命令`, ``];
			for (const [name, cmd] of registry.getAllCommands()) {
				if (isGroup && cmd.c2cOnly) continue;
				lines.push(`<qqbot-cmd-input text="/${name}" show="/${name}"/> ${cmd.description}`);
			}
			lines.push(``, `> 插件版本 v${getPluginVersionString()}`);
			return lines.join("\n");
		}
	});
	registry.register({
		name: "bot-me",
		description: "查看当前发送者的账号ID",
		c2cOnly: true,
		usage: [
			`/bot-me`,
			``,
			`显示当前发送者的账号ID`
		].join("\n"),
		handler: (ctx) => {
			return `你的账号ID：\`${ctx.senderId}\``;
		}
	});
	registry.register({
		name: "bot-ping",
		description: "测试 OpenClaw 与 QQ 之间的网络延迟",
		usage: [
			`/bot-ping`,
			``,
			`测试当前 OpenClaw 宿主机与 QQ 服务器之间的网络延迟。`,
			`返回网络传输耗时和插件处理耗时。`
		].join("\n"),
		handler: (ctx) => {
			const now = Date.now();
			const eventTime = new Date(ctx.eventTimestamp).getTime();
			if (Number.isNaN(eventTime)) return `✅ pong!`;
			const totalMs = now - eventTime;
			const qqToPlugin = ctx.receivedAt - eventTime;
			const pluginProcess = now - ctx.receivedAt;
			return [
				`✅ pong!`,
				``,
				`⏱ 延迟：${totalMs}ms`,
				`  ├ 网络传输：${qqToPlugin}ms`,
				`  └ 插件处理：${pluginProcess}ms`
			].join("\n");
		}
	});
	registry.register({
		name: "bot-version",
		description: "查看 QQBot 插件版本和 OpenClaw 框架版本",
		c2cOnly: true,
		usage: [
			`/bot-version`,
			``,
			`查看当前 QQBot 插件版本和 OpenClaw 框架版本。`
		].join("\n"),
		handler: async () => {
			const frameworkVersion = resolveRuntimeServiceVersion();
			const ver = getPluginVersionString();
			return [
				`🦞 OpenClaw 框架版本：${frameworkVersion}`,
				`🤖 QQBot 插件版本：v${ver}`,
				`🌟 官方 GitHub 仓库：[点击前往](${QQBOT_PLUGIN_GITHUB_URL})`
			].join("\n");
		}
	});
	registry.register({
		name: "bot-upgrade",
		description: "查看 QQBot 升级指引",
		c2cOnly: true,
		usage: [
			`/bot-upgrade`,
			``,
			`查看 QQBot 升级说明。`
		].join("\n"),
		handler: () => [`📘 QQBot 升级指引：`, `[点击查看升级说明](${QQBOT_UPGRADE_GUIDE_URL})`].join("\n")
	});
}
//#endregion
//#region extensions/qqbot/src/engine/commands/builtin/register-clear-storage.ts
function scanDirectoryFiles(dirPath) {
	const files = [];
	if (!fs.existsSync(dirPath)) return files;
	const walk = (dir) => {
		let entries;
		try {
			entries = fs.readdirSync(dir, { withFileTypes: true });
		} catch {
			return;
		}
		for (const entry of entries) {
			const fullPath = path.join(dir, entry.name);
			if (entry.isDirectory()) walk(fullPath);
			else if (entry.isFile()) try {
				const stat = fs.statSync(fullPath);
				files.push({
					filePath: fullPath,
					size: stat.size
				});
			} catch {}
		}
	};
	walk(dirPath);
	files.sort((a, b) => b.size - a.size);
	return files;
}
function formatBytes(bytes) {
	return formatByteSize(bytes, {
		style: "legacy-binary",
		maxUnit: "giga",
		separator: " ",
		fractionDigits: (_value, unit) => unit === "byte" ? null : 1
	});
}
function removeEmptyDirs(dirPath) {
	if (!fs.existsSync(dirPath)) return;
	let entries;
	try {
		entries = fs.readdirSync(dirPath, { withFileTypes: true });
	} catch {
		return;
	}
	for (const entry of entries) if (entry.isDirectory()) removeEmptyDirs(path.join(dirPath, entry.name));
	try {
		if (fs.readdirSync(dirPath).length === 0) fs.rmdirSync(dirPath);
	} catch {}
}
const CLEAR_STORAGE_MAX_DISPLAY = 10;
/**
* Resolve the canonical QQBot downloads directory.
*
* All inbound attachments and outbound fallback downloads are stored directly
* under `~/.openclaw/media/qqbot/downloads/` without appId subdivision.
* The clear-storage command therefore cleans the entire downloads root.
*/
function resolveQqbotDownloadsDir() {
	return getQQBotMediaPath("downloads");
}
function registerClearStorageCommands(registry) {
	registry.register({
		name: "bot-clear-storage",
		description: "清理通过 QQBot 对话产生的下载文件，释放主机磁盘空间",
		requireAuth: true,
		c2cOnly: true,
		usage: [
			`/bot-clear-storage`,
			``,
			`扫描 QQBot 下载目录下的所有文件并列出明细。`,
			`确认后执行删除，释放主机磁盘空间。`,
			``,
			`/bot-clear-storage --force   确认执行清理`,
			``,
			`⚠️ 仅在私聊中可用。`
		].join("\n"),
		handler: (ctx) => {
			const isForce = ctx.args.trim() === "--force";
			const targetDir = resolveQqbotDownloadsDir();
			const displayDir = `~/.openclaw/media/qqbot/downloads`;
			if (!isForce) {
				const files = scanDirectoryFiles(targetDir);
				if (files.length === 0) return [
					`✅ 当前没有需要清理的文件`,
					``,
					`目录 \`${displayDir}\` 为空或不存在。`
				].join("\n");
				const totalSize = files.reduce((sum, f) => sum + f.size, 0);
				const lines = [
					`即将清理 \`${displayDir}\` 目录下所有文件，总共 ${files.length} 个文件，占用磁盘存储空间 ${formatBytes(totalSize)}。`,
					``,
					`目录文件概况：`
				];
				const displayFiles = files.slice(0, CLEAR_STORAGE_MAX_DISPLAY);
				for (const f of displayFiles) {
					const relativePath = path.relative(targetDir, f.filePath).replace(/\\/g, "/");
					lines.push(`${relativePath} (${formatBytes(f.size)})`, ``, ``);
				}
				if (files.length > CLEAR_STORAGE_MAX_DISPLAY) lines.push(`...[合计：${files.length} 个文件（${formatBytes(totalSize)}）]`, ``);
				lines.push(``, `---`, ``, `确认清理后，上述保存在 OpenClaw 运行主机磁盘上的文件将永久删除，后续对话过程中 AI 无法再找回相关文件。`, `‼️ 点击指令确认删除`, `<qqbot-cmd-enter text="/bot-clear-storage --force" />`);
				return lines.join("\n");
			}
			const files = scanDirectoryFiles(targetDir);
			if (files.length === 0) return `✅ 目录已为空，无需清理`;
			let deletedCount = 0;
			let deletedSize = 0;
			let failedCount = 0;
			for (const f of files) try {
				fs.unlinkSync(f.filePath);
				deletedCount++;
				deletedSize += f.size;
			} catch {
				failedCount++;
			}
			try {
				removeEmptyDirs(targetDir);
			} catch {}
			if (failedCount === 0) return [
				`✅ 清理成功`,
				``,
				`已删除 ${deletedCount} 个文件，释放 ${formatBytes(deletedSize)} 磁盘空间。`
			].join("\n");
			return [
				`⚠️ 部分清理完成`,
				``,
				`已删除 ${deletedCount} 个文件（${formatBytes(deletedSize)}），${failedCount} 个文件删除失败。`
			].join("\n");
		}
	});
}
//#endregion
//#region extensions/qqbot/src/engine/commands/builtin/register-group-allways.ts
function registerGroupAllwaysCommand(registry) {
	registry.register({
		name: "bot-group-allways",
		description: "修改群消息默认响应模式",
		requireAuth: true,
		c2cOnly: true,
		usage: [
			`/bot-group-allways on   AI 自主判断何时发言（无需 @）`,
			`/bot-group-allways off  仅在被 @ 时回复`,
			`/bot-group-allways      查看当前设置`,
			``,
			`设为 on 后，AI 会自主判断每条消息是否需要回复（无需 @）。`,
			`仍可通过 groups.{groupId}.requireMention 对单个群覆盖。`,
			``,
			`优先级：具体群配置 > 通配符 "*" > defaultRequireMention（本指令）> 默认 true`
		].join("\n"),
		handler: async (ctx) => {
			const arg = ctx.args.trim().toLowerCase();
			const currentRequireMention = ctx.accountConfig?.defaultRequireMention ?? true;
			if (!arg) return [
				`🤖 群自主发言状态：${currentRequireMention ? "❌ 仅被 @ 时回复" : "✅ 自主判断何时发言"}`,
				`使用 <qqbot-cmd-input text="/bot-group-allways on" show="/bot-group-allways on"/> 设为自主发言`,
				`使用 <qqbot-cmd-input text="/bot-group-allways off" show="/bot-group-allways off"/> 设为仅被 @ 时回复`
			].join("\n");
			if (arg !== "on" && arg !== "off") return `❌ 参数错误，请使用 on 或 off\n\n示例：/bot-group-allways on`;
			const newRequireMention = arg === "off";
			if (newRequireMention === currentRequireMention) return `🤖 群自主发言已经是"${arg}"状态，无需操作`;
			let runtime;
			try {
				const getter = getApproveRuntimeGetter();
				if (!getter) throw new Error("runtime not available");
				runtime = getter();
			} catch {
				const fwVer = resolveRuntimeServiceVersion();
				const ver = getPluginVersionString();
				return [
					`❌ 当前版本不支持该指令`,
					``,
					`🦞框架版本：${fwVer}`,
					`🤖QQBot 插件版本：v${ver}`,
					``,
					`可通过以下命令手动设置：`,
					``,
					`\`\`\`shell`,
					`# 设为 AI 自主判断何时发言（defaultRequireMention=false）`,
					`openclaw config set channels.qqbot.defaultRequireMention false`,
					`# 或设为仅被 @ 时回复（defaultRequireMention=true）`,
					`openclaw config set channels.qqbot.defaultRequireMention true`,
					`\`\`\``
				].join("\n");
			}
			try {
				const configApi = runtime.config;
				const currentCfg = structuredClone(configApi.current());
				const qqbot = (currentCfg.channels ?? {}).qqbot;
				if (!qqbot) return `❌ 配置文件中未找到 qqbot 通道配置`;
				const accountId = ctx.accountId;
				if (accountId !== "default" && Boolean(qqbot.accounts?.[accountId])) {
					const nextAccounts = { ...qqbot.accounts ?? {} };
					const acct = { ...nextAccounts[accountId] };
					acct.defaultRequireMention = newRequireMention;
					nextAccounts[accountId] = acct;
					qqbot.accounts = nextAccounts;
				} else qqbot.defaultRequireMention = newRequireMention;
				await configApi.replaceConfigFile({
					nextConfig: currentCfg,
					afterWrite: { mode: "auto" }
				});
				clearRuntimeConfigSnapshot();
				return [
					`✅ 群自主发言已设置为 ${newRequireMention ? "**off**（仅被 @ 时回复）" : "**on**（AI 自主判断何时发言）"}`,
					``,
					newRequireMention ? `仅在被 @ 机器人才会回复。` : `AI 将自主判断群消息是否需要回复，无需被 @ 即可发言。`
				].join("\n");
			} catch (err) {
				return `❌ 配置写入失败: ${err instanceof Error ? err.message : String(err)}`;
			}
		}
	});
}
//#endregion
//#region extensions/qqbot/src/engine/commands/builtin/log-helpers.ts
/** Read user-configured log file paths from local config files. */
function getConfiguredLogFiles() {
	const homeDir = getHomeDir();
	const files = [];
	for (const cli of [
		"openclaw",
		"clawdbot",
		"moltbot"
	]) try {
		const logFile = loadJsonFile(path.join(homeDir, `.${cli}`, `${cli}.json`))?.logging?.file;
		if (logFile && typeof logFile === "string") files.push(path.resolve(logFile));
		break;
	} catch {}
	return files;
}
/** Collect directories that may contain runtime logs across common install layouts. */
function collectCandidateLogDirs() {
	const homeDir = getHomeDir();
	const dirs = /* @__PURE__ */ new Set();
	const pushDir = (p) => {
		if (!p) return;
		const normalized = path.resolve(p);
		dirs.add(normalized);
	};
	const pushStateDir = (stateDir) => {
		if (!stateDir) return;
		pushDir(stateDir);
		pushDir(path.join(stateDir, "logs"));
	};
	for (const logFile of getConfiguredLogFiles()) pushDir(path.dirname(logFile));
	for (const [key, value] of Object.entries(process.env)) {
		if (!value) continue;
		if (/STATE_DIR$/i.test(key) && /(OPENCLAW|CLAWDBOT|MOLTBOT)/i.test(key)) pushStateDir(value);
	}
	for (const name of [
		".openclaw",
		".clawdbot",
		".moltbot",
		"openclaw",
		"clawdbot",
		"moltbot"
	]) {
		pushDir(path.join(homeDir, name));
		pushDir(path.join(homeDir, name, "logs"));
	}
	const searchRoots = /* @__PURE__ */ new Set([
		homeDir,
		process.cwd(),
		path.dirname(process.cwd())
	]);
	if (process.env.APPDATA) searchRoots.add(process.env.APPDATA);
	if (process.env.LOCALAPPDATA) searchRoots.add(process.env.LOCALAPPDATA);
	for (const root of searchRoots) try {
		const entries = fs.readdirSync(root, { withFileTypes: true });
		for (const entry of entries) {
			if (!entry.isDirectory()) continue;
			if (!/(openclaw|clawdbot|moltbot)/i.test(entry.name)) continue;
			const base = path.join(root, entry.name);
			pushDir(base);
			pushDir(path.join(base, "logs"));
		}
	} catch {}
	if (!isWindows()) for (const name of [
		"openclaw",
		"clawdbot",
		"moltbot"
	]) pushDir(path.join("/var/log", name));
	const tmpRoots = /* @__PURE__ */ new Set();
	if (isWindows()) {
		tmpRoots.add("C:\\tmp");
		if (process.env.TEMP) tmpRoots.add(process.env.TEMP);
		if (process.env.TMP) tmpRoots.add(process.env.TMP);
		if (process.env.LOCALAPPDATA) tmpRoots.add(path.join(process.env.LOCALAPPDATA, "Temp"));
	} else tmpRoots.add("/tmp");
	for (const tmpRoot of tmpRoots) for (const name of [
		"openclaw",
		"clawdbot",
		"moltbot"
	]) pushDir(path.join(tmpRoot, name));
	return Array.from(dirs);
}
function addCollisionSuffix(filePath, suffix) {
	const ext = path.extname(filePath);
	const baseName = path.basename(filePath, ext);
	return path.join(path.dirname(filePath), `${baseName}-${suffix}${ext}`);
}
function writeNewTextFileSync(filePath, contents) {
	for (let suffix = 1; suffix <= 100; suffix++) {
		const candidate = suffix === 1 ? filePath : addCollisionSuffix(filePath, suffix);
		try {
			fs.writeFileSync(candidate, contents, {
				encoding: "utf8",
				flag: "wx"
			});
			return candidate;
		} catch (error) {
			if (typeof error === "object" && error && "code" in error && error.code === "EEXIST") continue;
			throw error;
		}
	}
	throw new Error(`Could not find an unused log export filename near ${filePath}`);
}
function collectRecentLogFiles(logDirs) {
	const candidates = [];
	const dedupe = /* @__PURE__ */ new Set();
	const pushFile = (filePath, sourceDir) => {
		const normalized = path.resolve(filePath);
		if (dedupe.has(normalized)) return;
		try {
			const stat = fs.statSync(normalized);
			if (!stat.isFile()) return;
			dedupe.add(normalized);
			candidates.push({
				filePath: normalized,
				sourceDir,
				mtimeMs: stat.mtimeMs
			});
		} catch {}
	};
	for (const logFile of getConfiguredLogFiles()) pushFile(logFile, path.dirname(logFile));
	for (const dir of logDirs) {
		pushFile(path.join(dir, "gateway.log"), dir);
		pushFile(path.join(dir, "gateway.err.log"), dir);
		pushFile(path.join(dir, "openclaw.log"), dir);
		pushFile(path.join(dir, "clawdbot.log"), dir);
		pushFile(path.join(dir, "moltbot.log"), dir);
		try {
			const entries = fs.readdirSync(dir, { withFileTypes: true });
			for (const entry of entries) {
				if (!entry.isFile()) continue;
				if (!/\.(log|txt)$/i.test(entry.name)) continue;
				if (!/(gateway|openclaw|clawdbot|moltbot)/i.test(entry.name)) continue;
				pushFile(path.join(dir, entry.name), dir);
			}
		} catch {}
	}
	candidates.sort((a, b) => b.mtimeMs - a.mtimeMs);
	return candidates;
}
/**
* Read the last N lines of a file without loading the entire file into memory.
*/
function tailFileLines(filePath, maxLines) {
	const fd = fs.openSync(filePath, "r");
	try {
		const fileSize = fs.fstatSync(fd).size;
		if (fileSize === 0) return {
			tail: [],
			totalFileLines: 0
		};
		const CHUNK_SIZE = 64 * 1024;
		const chunks = [];
		let bytesRead = 0;
		let position = fileSize;
		let newlineCount = 0;
		while (position > 0 && newlineCount <= maxLines) {
			const readSize = Math.min(CHUNK_SIZE, position);
			position -= readSize;
			const buf = Buffer.alloc(readSize);
			fs.readSync(fd, buf, 0, readSize, position);
			chunks.unshift(buf);
			bytesRead += readSize;
			for (let i = 0; i < readSize; i++) if (buf[i] === 10) newlineCount++;
		}
		const allLines = Buffer.concat(chunks).toString("utf8").split("\n");
		const tail = allLines.slice(-maxLines);
		let totalFileLines;
		if (bytesRead >= fileSize) totalFileLines = allLines.length;
		else {
			const avgBytesPerLine = bytesRead / Math.max(allLines.length, 1);
			totalFileLines = Math.round(fileSize / avgBytesPerLine);
		}
		return {
			tail,
			totalFileLines
		};
	} finally {
		fs.closeSync(fd);
	}
}
/**
* Build the /bot-logs result: collect recent log files, write them to a temp file.
*/
function buildBotLogsResult() {
	const logDirs = collectCandidateLogDirs();
	const recentFiles = collectRecentLogFiles(logDirs).slice(0, 4);
	if (recentFiles.length === 0) {
		const existingDirs = logDirs.filter((d) => {
			try {
				return fs.existsSync(d);
			} catch {
				return false;
			}
		});
		const searched = existingDirs.length > 0 ? existingDirs.map((d) => `  • ${d}`).join("\n") : logDirs.slice(0, 6).map((d) => `  • ${d}`).join("\n") + (logDirs.length > 6 ? `\n  …以及另外 ${logDirs.length - 6} 个路径` : "");
		return [
			`⚠️ 未找到日志文件`,
			``,
			`已搜索以下${existingDirs.length > 0 ? "存在的" : ""}路径：`,
			searched,
			``,
			`💡 如果日志存放在自定义路径，请在配置中添加：`,
			`  "logging": { "file": "/path/to/your/logfile.log" }`
		].join("\n");
	}
	const lines = [];
	let totalIncluded = 0;
	let totalOriginal = 0;
	let truncatedCount = 0;
	const MAX_LINES_PER_FILE = 1e3;
	for (const logFile of recentFiles) try {
		const { tail, totalFileLines } = tailFileLines(logFile.filePath, MAX_LINES_PER_FILE);
		if (tail.length > 0) {
			const fileName = path.basename(logFile.filePath);
			lines.push(`\n========== ${fileName} (last ${tail.length} of ${totalFileLines} lines) ==========`);
			lines.push(`from: ${logFile.sourceDir}`);
			lines.push(...tail);
			totalIncluded += tail.length;
			totalOriginal += totalFileLines;
			if (totalFileLines > MAX_LINES_PER_FILE) truncatedCount++;
		}
	} catch {
		lines.push(`[Failed to read ${path.basename(logFile.filePath)}]`);
	}
	if (lines.length === 0) return `⚠️ 找到了日志文件，但无法读取。请检查文件权限。`;
	const tmpDir = getQQBotDataDir("downloads");
	const timestamp = (/* @__PURE__ */ new Date()).toISOString().replace(/[:.]/g, "-").slice(0, 19);
	const tmpFile = writeNewTextFileSync(path.join(tmpDir, `bot-logs-${timestamp}.txt`), lines.join("\n"));
	const fileCount = recentFiles.length;
	const topSources = uniqueStrings(recentFiles.map((item) => item.sourceDir)).slice(0, 3);
	let summaryText = `共 ${fileCount} 个日志文件，包含 ${totalIncluded} 行内容`;
	if (truncatedCount > 0) summaryText += `（其中 ${truncatedCount} 个文件已截断为最后 ${MAX_LINES_PER_FILE} 行，总计原始 ${totalOriginal} 行）`;
	return {
		text: `📋 ${summaryText}\n📂 来源：${topSources.join(" | ")}`,
		filePath: tmpFile
	};
}
//#endregion
//#region extensions/qqbot/src/engine/commands/builtin/register-logs.ts
function registerLogCommands(registry) {
	registry.register({
		name: "bot-logs",
		description: "导出本地日志文件",
		requireAuth: true,
		c2cOnly: true,
		usage: [
			`/bot-logs`,
			``,
			`导出最近的 OpenClaw 日志文件（最多 4 个文件）。`,
			`每个文件只保留最后 1000 行，并作为附件返回。`
		].join("\n"),
		handler: () => {
			return buildBotLogsResult();
		}
	});
}
//#endregion
//#region extensions/qqbot/src/engine/commands/builtin/register-streaming.ts
function isStreamingConfigEnabled(streaming) {
	if (streaming === true) return true;
	if (streaming === false || streaming === void 0 || streaming === null) return false;
	if (typeof streaming === "object") {
		const o = streaming;
		if (o.c2cStreamApi === true) return true;
		if (o.mode === "off") return false;
		return true;
	}
	return false;
}
function registerStreamingCommands(registry) {
	registry.register({
		name: "bot-streaming",
		description: "一键开关流式消息",
		requireAuth: true,
		c2cOnly: true,
		usage: [
			`/bot-streaming on     开启流式消息`,
			`/bot-streaming off    关闭流式消息`,
			`/bot-streaming        查看当前流式消息状态`,
			``,
			`开启后，AI 的回复会以流式形式逐步显示（打字机效果）。`,
			`注意：仅 C2C（私聊）支持流式消息。`
		].join("\n"),
		handler: async (ctx) => {
			const arg = ctx.args.trim().toLowerCase();
			const currentOn = isStreamingConfigEnabled(ctx.accountConfig?.streaming);
			if (!arg) return [
				`📡 流式消息状态：${currentOn ? "✅ 已开启" : "❌ 已关闭"}`,
				``,
				`使用 <qqbot-cmd-input text="/bot-streaming on" show="/bot-streaming on"/> 开启`,
				`使用 <qqbot-cmd-input text="/bot-streaming off" show="/bot-streaming off"/> 关闭`
			].join("\n");
			if (arg !== "on" && arg !== "off") return `❌ 参数错误，请使用 on 或 off\n\n示例：/bot-streaming on`;
			const wantOn = arg === "on";
			if (wantOn === currentOn) return `📡 流式消息已经是${wantOn ? "开启" : "关闭"}状态，无需操作`;
			let runtime;
			try {
				const getter = getApproveRuntimeGetter();
				if (!getter) throw new Error("runtime not available");
				runtime = getter();
			} catch {
				const fwVer = resolveRuntimeServiceVersion();
				const ver = getPluginVersionString();
				return [
					`❌ 当前版本不支持该指令`,
					``,
					`🦞框架版本：${fwVer}`,
					`🤖QQBot 插件版本：v${ver}`,
					``,
					`可通过以下命令手动开启流式消息：`,
					``,
					`\`\`\`shell`,
					`# 1. 开启流式消息`,
					`openclaw config set channels.qqbot.streaming true`,
					``,
					`# 2. 重启网关使配置生效`,
					`openclaw gateway restart`,
					`\`\`\``
				].join("\n");
			}
			try {
				const configApi = runtime.config;
				const currentCfg = structuredClone(configApi.current());
				const qqbot = (currentCfg.channels ?? {}).qqbot;
				if (!qqbot) return `❌ 配置文件中未找到 qqbot 通道配置`;
				const accountId = ctx.accountId;
				const newVal = wantOn;
				if (accountId !== "default") {
					const nextAccounts = { ...qqbot.accounts ?? {} };
					const acct = { ...nextAccounts[accountId] };
					acct.streaming = newVal;
					nextAccounts[accountId] = acct;
					qqbot.accounts = nextAccounts;
				} else {
					qqbot.streaming = newVal;
					const accs = qqbot.accounts;
					if (accs?.default && typeof accs.default === "object") {
						const nextAccs = { ...accs };
						nextAccs.default = {
							...accs.default,
							streaming: newVal
						};
						qqbot.accounts = nextAccs;
					}
				}
				await configApi.replaceConfigFile({
					nextConfig: currentCfg,
					afterWrite: { mode: "auto" }
				});
				return [
					`✅ 流式消息已${wantOn ? "开启" : "关闭"}`,
					``,
					wantOn ? `AI 的回复将以流式形式逐步显示（仅私聊生效）。` : `AI 的回复将恢复为完整发送。`
				].join("\n");
			} catch (err) {
				return `❌ 配置写入失败: ${err instanceof Error ? err.message : String(err)}`;
			}
		}
	});
}
//#endregion
//#region extensions/qqbot/src/engine/commands/builtin/register-all.ts
/**
* Register all built-in slash commands on the shared registry instance.
*/
function registerBuiltinSlashCommands(registry) {
	registerBasicBotCommands(registry);
	registerLogCommands(registry);
	registerClearStorageCommands(registry);
	registerStreamingCommands(registry);
	registerApproveCommands(registry);
	registerGroupAllwaysCommand(registry);
}
//#endregion
//#region extensions/qqbot/src/engine/commands/command-visibility.ts
const PRIVATE_CHAT_ONLY_TEXT = "该命令仅限私聊使用，请在私聊中发送。";
const GROUP_VISIBLE_CORE_COMMANDS = /* @__PURE__ */ new Set([
	"help",
	"btw",
	"side",
	"stop"
]);
const STRICT_CORE_COMMANDS = /* @__PURE__ */ new Set(["new", "reset"]);
const GROUP_HIDDEN_CORE_COMMANDS = /* @__PURE__ */ new Set([
	"goal",
	"usage",
	"activation",
	"send",
	"reset",
	"new",
	"name",
	"compact",
	"think",
	"thinking",
	"t",
	"fast",
	"reasoning",
	"reason",
	"queue"
]);
const PRIVATE_ONLY_CORE_COMMANDS = /* @__PURE__ */ new Set([
	"commands",
	"tools",
	"skill",
	"diagnostics",
	"crestodian",
	"tasks",
	"allowlist",
	"approve",
	"context",
	"export-session",
	"export",
	"export-trajectory",
	"trajectory",
	"tts",
	"whoami",
	"id",
	"session",
	"subagents",
	"acp",
	"focus",
	"unfocus",
	"agents",
	"steer",
	"tell",
	"config",
	"mcp",
	"plugins",
	"plugin",
	"debug",
	"status",
	"restart",
	"trace",
	"verbose",
	"v",
	"elevated",
	"elev",
	"exec",
	"model",
	"models",
	"bash"
]);
function parseSlashCommandName(content) {
	const trimmed = (content ?? "").trim();
	if (!trimmed.startsWith("/")) return;
	return ((trimmed.slice(1).split(/\s+/, 1)[0]?.trim().toLowerCase() ?? "").split(/[@:：]/u, 1)[0] ?? "") || void 0;
}
function classifyCoreCommandForGroup(content, commandLevel = "all") {
	const commandName = parseSlashCommandName(content);
	if (!commandName) return { visibility: "unknown" };
	if (commandLevel === "all") return {
		commandName,
		visibility: GROUP_VISIBLE_CORE_COMMANDS.has(commandName) ? "group" : "hidden"
	};
	if (commandLevel === "strict") {
		if (commandName === "stop") return {
			commandName,
			visibility: "group"
		};
		if (STRICT_CORE_COMMANDS.has(commandName)) return {
			commandName,
			visibility: "hidden"
		};
		return {
			commandName,
			visibility: "private"
		};
	}
	if (GROUP_VISIBLE_CORE_COMMANDS.has(commandName)) return {
		commandName,
		visibility: "group"
	};
	if (GROUP_HIDDEN_CORE_COMMANDS.has(commandName)) return {
		commandName,
		visibility: "hidden"
	};
	if (PRIVATE_ONLY_CORE_COMMANDS.has(commandName)) return {
		commandName,
		visibility: "private"
	};
	return {
		commandName,
		visibility: "unknown"
	};
}
//#endregion
//#region extensions/qqbot/src/engine/commands/slash-commands.ts
/** Lowercase and trim a string. */
function lc(s) {
	return (s ?? "").toLowerCase().trim();
}
/**
* Slash command registry.
*
* Maintains two maps:
* - `commands` — QQBot message-flow commands
* - `frameworkCommands` — auth-gated commands that are safe on the framework surface
*/
var SlashCommandRegistry = class {
	constructor() {
		this.commands = /* @__PURE__ */ new Map();
		this.frameworkCommands = /* @__PURE__ */ new Map();
	}
	/** Register one command. */
	register(cmd) {
		const key = lc(cmd.name);
		this.commands.set(key, cmd);
		if (cmd.requireAuth) this.frameworkCommands.set(key, cmd);
	}
	/** Return all commands that may be registered on the framework surface. */
	getFrameworkCommands() {
		return Array.from(this.frameworkCommands.values()).map((cmd) => ({
			name: cmd.name,
			description: cmd.description,
			usage: cmd.usage,
			c2cOnly: cmd.c2cOnly,
			handler: cmd.handler
		}));
	}
	/** Return all registered commands (both maps) for help listing. */
	getAllCommands() {
		const all = /* @__PURE__ */ new Map();
		for (const [k, v] of this.commands) all.set(k, v);
		for (const [k, v] of this.frameworkCommands) all.set(k, v);
		return all;
	}
	/**
	* Try to match and execute a pre-dispatch slash command.
	*
	* @returns A reply when matched, or null when the message should continue
	*          through normal routing.
	*/
	async matchSlashCommand(ctx, log) {
		const content = ctx.rawContent.trim();
		if (!content.startsWith("/")) return null;
		const spaceIdx = content.indexOf(" ");
		const cmdName = lc(spaceIdx === -1 ? content.slice(1) : content.slice(1, spaceIdx));
		const args = spaceIdx === -1 ? "" : content.slice(spaceIdx + 1).trim();
		const cmd = this.commands.get(cmdName);
		if (!cmd) return null;
		const isGroup = ctx.type === "group" || ctx.type === "guild";
		const groupCommandLevel = ctx.groupCommandLevel ?? "all";
		if (isGroup && groupCommandLevel === "strict") return PRIVATE_CHAT_ONLY_TEXT;
		if (cmd.c2cOnly && ctx.type !== "c2c") return PRIVATE_CHAT_ONLY_TEXT;
		if (cmd.requireAuth && !ctx.commandAuthorized) {
			log?.info?.(`[qqbot] Slash command /${cmd.name} rejected: sender ${ctx.senderId} is not authorized`);
			return `⛔ 权限不足：请先在 channels.qqbot.${isGroup ? "groupAllowFrom" : "allowFrom"} 中配置明确的发送者列表后再使用 /${cmd.name}。`;
		}
		if (args === "?") {
			if (cmd.usage) return `📖 /${cmd.name} 用法：\n\n${cmd.usage}`;
			return `/${cmd.name} - ${cmd.description}`;
		}
		ctx.args = args;
		return await cmd.handler(ctx);
	}
};
//#endregion
//#region extensions/qqbot/src/engine/commands/slash-commands-impl.ts
const registry = new SlashCommandRegistry();
registerBuiltinSlashCommands(registry);
/**
* Initialize command dependencies from the EngineAdapters.commands port.
* Called once by the bridge layer during startup.
*/
function initCommands(port) {
	initSlashCommandDeps(port);
}
/**
* Return commands that may be registered with the framework via
* api.registerCommand() in registerFull().
*/
function getFrameworkCommands() {
	return registry.getFrameworkCommands();
}
/**
* Try to match and execute a plugin-level slash command.
*
* @returns A reply when matched, or null when the message should continue through normal routing.
*/
async function matchSlashCommand(ctx) {
	return registry.matchSlashCommand(ctx, { info: debugLog });
}
/** Return the plugin version for external callers. */
function getPluginVersion() {
	return getPluginVersionString();
}
/** Return the framework version for external callers. */
function getFrameworkVersion() {
	return getFrameworkVersionString();
}
const DEFAULT_GROUP_PROMPT = "If the sender is a bot, respond only when they explicitly @mention you to ask a question or request assistance with a specific task; keep your replies concise and clear, avoiding the urge to race other bots to answer or engage in lengthy, unproductive exchanges. In group chats, prioritize responding to messages from human users; bots should maintain a collaborative rather than competitive dynamic to ensure the conversation remains orderly and does not result in message flooding.";
const DEFAULT_GROUP_CONFIG = {
	requireMention: true,
	ignoreOtherMentions: false,
	commandLevel: "all",
	name: "",
	historyLimit: 50
};
function readGroupsMap(cfg, accountId) {
	const groups = asOptionalObjectRecord(resolveAccountBase(cfg, accountId).config.groups);
	if (!groups) return {};
	const normalized = {};
	for (const [key, value] of Object.entries(groups)) {
		const sub = asOptionalObjectRecord(value);
		if (sub) normalized[key] = sub;
	}
	return normalized;
}
function readBoolean(obj, key) {
	return asBoolean(obj[key]);
}
function readString(obj, key) {
	const v = obj[key];
	return typeof v === "string" && v.length > 0 ? v : void 0;
}
function readCommandLevel(obj, key) {
	const v = readString(obj, key);
	return v === "all" || v === "safety" || v === "strict" ? v : void 0;
}
function readHistoryLimit(obj, key) {
	const v = obj[key];
	if (typeof v !== "number" || !Number.isFinite(v)) return;
	return Math.max(0, Math.floor(v));
}
function resolveGroupConfig(cfg, groupOpenid, accountId) {
	const account = resolveAccountBase(cfg, accountId);
	const groups = readGroupsMap(cfg, accountId);
	const wildcard = groups["*"] ?? {};
	const specific = groupOpenid ? groups[groupOpenid] ?? {} : {};
	const accountDefaultRequireMention = asBoolean(account.config.defaultRequireMention) ?? DEFAULT_GROUP_CONFIG.requireMention;
	return {
		requireMention: readBoolean(specific, "requireMention") ?? readBoolean(wildcard, "requireMention") ?? accountDefaultRequireMention,
		ignoreOtherMentions: readBoolean(specific, "ignoreOtherMentions") ?? readBoolean(wildcard, "ignoreOtherMentions") ?? DEFAULT_GROUP_CONFIG.ignoreOtherMentions,
		commandLevel: readCommandLevel(specific, "commandLevel") ?? readCommandLevel(wildcard, "commandLevel") ?? DEFAULT_GROUP_CONFIG.commandLevel,
		name: readString(specific, "name") ?? readString(wildcard, "name") ?? DEFAULT_GROUP_CONFIG.name,
		prompt: readString(specific, "prompt") ?? readString(wildcard, "prompt"),
		historyLimit: readHistoryLimit(specific, "historyLimit") ?? readHistoryLimit(wildcard, "historyLimit") ?? DEFAULT_GROUP_CONFIG.historyLimit
	};
}
function resolveGroupCommandLevelFromAccountConfig(accountConfig, groupOpenid) {
	const groups = asOptionalObjectRecord(accountConfig?.groups);
	const wildcard = asOptionalObjectRecord(groups?.["*"]) ?? {};
	return readCommandLevel(groupOpenid ? asOptionalObjectRecord(groups?.[groupOpenid]) ?? {} : {}, "commandLevel") ?? readCommandLevel(wildcard, "commandLevel") ?? DEFAULT_GROUP_CONFIG.commandLevel;
}
function resolveGroupSettings(params) {
	const config = resolveGroupConfig(params.cfg, params.groupOpenid, params.accountId);
	return {
		config,
		name: config.name || params.groupOpenid.slice(0, 8),
		mentionPatterns: resolveMentionPatterns(params.cfg, params.agentId)
	};
}
function resolveMentionPatterns(cfg, agentId) {
	if (agentId) {
		const agents = asOptionalObjectRecord(cfg.agents);
		const agentGroupChat = (Array.isArray(agents?.list) ? agents?.list : []).find((a) => typeof a.id === "string" && a.id.trim().toLowerCase() === agentId.trim().toLowerCase())?.groupChat;
		if (agentGroupChat && Object.hasOwn(agentGroupChat, "mentionPatterns")) {
			const patterns = agentGroupChat.mentionPatterns;
			return Array.isArray(patterns) ? patterns.filter((p) => typeof p === "string") : [];
		}
	}
	const globalGroupChat = asOptionalObjectRecord(asOptionalObjectRecord(cfg.messages)?.groupChat);
	if (globalGroupChat && Object.hasOwn(globalGroupChat, "mentionPatterns")) {
		const patterns = globalGroupChat.mentionPatterns;
		return Array.isArray(patterns) ? patterns.filter((p) => typeof p === "string") : [];
	}
	return [];
}
//#endregion
export { resolveMentionPatterns as a, getPluginVersion as c, PRIVATE_CHAT_ONLY_TEXT as d, classifyCoreCommandForGroup as f, resolveGroupSettings as i, initCommands as l, resolveGroupCommandLevelFromAccountConfig as n, getFrameworkCommands as o, resolveGroupConfig as r, getFrameworkVersion as s, DEFAULT_GROUP_PROMPT as t, matchSlashCommand as u };
