// @openclaw/qqbot/dist/log-DEtcoDWe.js
import { truncateUtf16Safe } from "openclaw/plugin-sdk/text-utility-runtime";
//#region extensions/qqbot/src/engine/utils/format.ts
/**
* General formatting and string utilities.
* 通用格式化与字符串工具。
*
* Pure utility functions with zero external dependencies.
* Replaces `openclaw/plugin-sdk/error-runtime` and `text-runtime`
* helpers for use inside engine/.
*
* NOTE: The framework `formatErrorMessage` also applies `redactSensitiveText()`
* for token masking. We intentionally omit that here — the framework's log
* pipeline handles redaction at a higher level.
*/
/**
* Format any error object into a readable string.
* 将任意错误对象格式化为可读字符串。
*
* Traverses the `.cause` chain for nested Error objects to include
* the full error context (e.g. network errors wrapped inside HTTP errors).
*/
function formatErrorMessage(err) {
	if (err instanceof Error) {
		let formatted = err.message || err.name || "Error";
		let cause = err.cause;
		const seen = /* @__PURE__ */ new Set([err]);
		while (cause && !seen.has(cause)) {
			seen.add(cause);
			if (cause instanceof Error) {
				if (cause.message) formatted += ` | ${cause.message}`;
				cause = cause.cause;
			} else if (typeof cause === "string") {
				formatted += ` | ${cause}`;
				break;
			} else break;
		}
		return formatted;
	}
	if (typeof err === "string") return err;
	if (err === null || err === void 0 || typeof err === "number" || typeof err === "boolean" || typeof err === "bigint") return String(err);
	try {
		return JSON.stringify(err);
	} catch {
		return Object.prototype.toString.call(err);
	}
}
/** Format a millisecond duration into a human-readable string (e.g. "5m 30s"). */
function formatDuration(durationMs) {
	const seconds = Math.round(durationMs / 1e3);
	if (seconds < 60) return `${seconds}s`;
	const minutes = Math.floor(seconds / 60);
	const remainSeconds = seconds % 60;
	return remainSeconds > 0 ? `${minutes}m ${remainSeconds}s` : `${minutes}m`;
}
//#endregion
//#region extensions/qqbot/src/engine/utils/log.ts
/**
* QQBot debug logging utilities.
* QQBot 调试日志工具。
*
* Only outputs when the QQBOT_DEBUG environment variable is set,
* preventing user message content from leaking in production logs.
*/
function isQqbotDebugEnabled() {
	const value = process.env.QQBOT_DEBUG;
	if (typeof value !== "string") return false;
	switch (value.trim().toLowerCase()) {
		case "1":
		case "on":
		case "true":
		case "yes": return true;
		default: return false;
	}
}
const isDebug = () => isQqbotDebugEnabled();
const MAX_LOG_VALUE_CHARS = 4096;
function sanitizeDebugLogValue(value) {
	let text;
	if (typeof value === "string") text = value;
	else if (value instanceof Error) text = value.stack || value.message;
	else try {
		text = JSON.stringify(value) ?? String(value);
	} catch {
		text = String(value);
	}
	const sanitized = text.replace(/\p{Cc}/gu, " ").replace(/\s+/g, " ").trim();
	if (sanitized.length <= MAX_LOG_VALUE_CHARS) return sanitized;
	return `${truncateUtf16Safe(sanitized, MAX_LOG_VALUE_CHARS)}...`;
}
function formatDebugLogArgs(args) {
	return args.map(sanitizeDebugLogValue).join(" ");
}
/** Debug-level log; only outputs when QQBOT_DEBUG is enabled. */
function debugLog(...args) {
	if (isDebug()) console.log(formatDebugLogArgs(args).replace(/\n|\r/g, ""));
}
/** Debug-level warning; only outputs when QQBOT_DEBUG is enabled. */
function debugWarn(...args) {
	if (isDebug()) console.warn(formatDebugLogArgs(args).replace(/\n|\r/g, ""));
}
/** Debug-level error; only outputs when QQBOT_DEBUG is enabled. */
function debugError(...args) {
	if (isDebug()) console.error(formatDebugLogArgs(args).replace(/\n|\r/g, ""));
}
//#endregion
export { formatErrorMessage as a, formatDuration as i, debugLog as n, debugWarn as r, debugError as t };
