// @openclaw/qqbot/dist/request-context-Bm7PTBD1.js
import { AsyncLocalStorage } from "node:async_hooks";
//#region extensions/qqbot/src/engine/utils/request-context.ts
/**
* Request-level context using AsyncLocalStorage.
*
* Provides ambient context (accountId, target openid, chat type, etc.)
* throughout the request lifecycle without explicit parameter threading.
*
* Gateway establishes the scope around each inbound message via
* `runWithRequestContext()`; any async code within that scope (including
* AI agent calls and tool `execute` callbacks) can retrieve the current
* request via `getRequestContext()` without racing with concurrent
* inbound messages.
*
* This is a pure Node.js module with zero framework dependencies,
* making it trivially portable between the built-in and standalone
* versions of QQBot.
*/
const store = new AsyncLocalStorage();
/**
* Execute an async function with request-scoped context.
*
* All code running within `fn` (including nested async calls) can
* retrieve the context via `getRequestContext()`.
*
* @param ctx - The context to attach to this request.
* @param fn - The async function to run within the context.
* @returns The return value of `fn`.
*/
function runWithRequestContext(ctx, fn) {
	return store.run(ctx, fn);
}
/**
* Retrieve the current request context.
*
* Returns `undefined` when called outside of a `runWithRequestContext`
* scope.
*/
function getRequestContext() {
	return store.getStore();
}
//#endregion
export { runWithRequestContext as n, getRequestContext as t };
