// @openclaw/qqbot/dist/state-keys-CQKlAFyo.js
import crypto from "node:crypto";
import { asObjectRecord } from "openclaw/plugin-sdk/runtime-doctor";
//#region extensions/qqbot/src/doctor-contract.ts
const RESTRICTED_GROUP_TOOLS = { deny: [
	"exec",
	"read",
	"write"
] };
function hasLegacyGroupToolPolicy(value) {
	const groups = asObjectRecord(value);
	if (!groups) return false;
	return Object.values(groups).some((group) => asObjectRecord(group)?.toolPolicy !== void 0);
}
function hasLegacyAccountGroupToolPolicy(value) {
	const accounts = asObjectRecord(value);
	if (!accounts) return false;
	return Object.values(accounts).some((account) => hasLegacyGroupToolPolicy(asObjectRecord(account)?.groups));
}
function migrateToolPolicy(value) {
	if (value === "none") return { deny: ["*"] };
	if (value === "full") return { allow: [] };
	if (value === "restricted") return { ...RESTRICTED_GROUP_TOOLS };
}
function describeToolPolicy(value) {
	return typeof value === "string" ? value : String(value);
}
function migrateGroups(params) {
	let changed = false;
	const nextGroups = { ...params.groups };
	for (const [groupId, rawGroup] of Object.entries(params.groups)) {
		const group = asObjectRecord(rawGroup);
		if (!group || group.toolPolicy === void 0) continue;
		const { toolPolicy, ...rest } = group;
		const nextGroup = { ...rest };
		const policy = migrateToolPolicy(toolPolicy);
		const path = `${params.pathPrefix}.${groupId}`;
		if (nextGroup.tools !== void 0) params.changes.push(`Removed ${path}.toolPolicy (${path}.tools already exists).`);
		else if (policy) {
			nextGroup.tools = policy;
			params.changes.push(`Moved ${path}.toolPolicy=${describeToolPolicy(toolPolicy)} to ${path}.tools.`);
		} else params.changes.push(`Removed unsupported ${path}.toolPolicy=${describeToolPolicy(toolPolicy)}.`);
		nextGroups[groupId] = nextGroup;
		changed = true;
	}
	return {
		groups: nextGroups,
		changed
	};
}
const legacyConfigRules = [{
	path: [
		"channels",
		"qqbot",
		"groups"
	],
	message: "channels.qqbot.groups.<id>.toolPolicy is legacy and was ignored by QQBot group tool enforcement; use channels.qqbot.groups.<id>.tools instead. Run \"openclaw doctor --fix\".",
	match: hasLegacyGroupToolPolicy
}, {
	path: [
		"channels",
		"qqbot",
		"accounts"
	],
	message: "channels.qqbot.accounts.<id>.groups.<groupId>.toolPolicy is legacy and was ignored by QQBot group tool enforcement; use channels.qqbot.accounts.<id>.groups.<groupId>.tools instead. Run \"openclaw doctor --fix\".",
	match: hasLegacyAccountGroupToolPolicy
}];
function normalizeCompatibilityConfig({ cfg }) {
	const rawEntry = asObjectRecord(cfg.channels?.qqbot);
	if (!rawEntry) return {
		config: cfg,
		changes: []
	};
	const changes = [];
	let updated = rawEntry;
	let changed = false;
	const groups = asObjectRecord(updated.groups);
	if (groups) {
		const migrated = migrateGroups({
			groups,
			pathPrefix: "channels.qqbot.groups",
			changes
		});
		if (migrated.changed) {
			updated = {
				...updated,
				groups: migrated.groups
			};
			changed = true;
		}
	}
	const accounts = asObjectRecord(updated.accounts);
	if (accounts) {
		let accountsChanged = false;
		const nextAccounts = { ...accounts };
		for (const [accountId, rawAccount] of Object.entries(accounts)) {
			const account = asObjectRecord(rawAccount);
			const accountGroups = asObjectRecord(account?.groups);
			if (!account || !accountGroups) continue;
			const migrated = migrateGroups({
				groups: accountGroups,
				pathPrefix: `channels.qqbot.accounts.${accountId}.groups`,
				changes
			});
			if (migrated.changed) {
				nextAccounts[accountId] = {
					...account,
					groups: migrated.groups
				};
				accountsChanged = true;
			}
		}
		if (accountsChanged) {
			updated = {
				...updated,
				accounts: nextAccounts
			};
			changed = true;
		}
	}
	if (!changed) return {
		config: cfg,
		changes: []
	};
	return {
		config: {
			...cfg,
			channels: {
				...cfg.channels,
				qqbot: updated
			}
		},
		changes
	};
}
//#endregion
//#region extensions/qqbot/src/engine/utils/state-keys.ts
function buildQQBotStateKey(...parts) {
	return crypto.createHash("sha256").update(JSON.stringify(parts)).digest("hex");
}
//#endregion
export { legacyConfigRules as n, normalizeCompatibilityConfig as r, buildQQBotStateKey as t };
