// @openclaw/qqbot/dist/string-normalize-R_0cKO7Q.js
//#region extensions/qqbot/src/engine/adapter/index.ts
let platformAdapter = null;
let platformAdapterFactory = null;
function registerPlatformAdapter(adapter) {
	platformAdapter = adapter;
}
function registerPlatformAdapterFactory(factory) {
	platformAdapterFactory = factory;
}
function getPlatformAdapter() {
	if (!platformAdapter && platformAdapterFactory) platformAdapter = platformAdapterFactory();
	if (!platformAdapter) throw new Error("PlatformAdapter not registered. Call registerPlatformAdapter() during bootstrap.");
	return platformAdapter;
}
function hasPlatformAdapter() {
	return platformAdapter !== null || platformAdapterFactory !== null;
}
//#endregion
//#region extensions/qqbot/src/engine/utils/string-normalize.ts
/**
* String normalization and record-coercion helpers.
*
* These are self-contained re-implementations of the functions that
* the plugin previously imported from broad SDK text barrels
* and shared record/string coercion helpers.
*
* core/ modules use these instead of importing plugin-sdk, keeping the
* shared layer portable between the built-in and standalone versions.
*/
/** Return the trimmed string or `null` when the value is not a non-empty string. */
function normalizeNullableString(value) {
	if (typeof value !== "string") return null;
	const trimmed = value.trim();
	return trimmed ? trimmed : null;
}
/** Return the trimmed string or `undefined` when the value is not a non-empty string. */
function normalizeOptionalString(value) {
	return normalizeNullableString(value) ?? void 0;
}
/**
* Stringify then normalize.  Accepts `string | number | boolean | bigint`.
* Returns `undefined` for objects, arrays, null, and undefined.
*/
function normalizeStringifiedOptionalString(value) {
	if (typeof value === "string") return normalizeOptionalString(value);
	if (typeof value === "number" || typeof value === "boolean" || typeof value === "bigint") return normalizeOptionalString(String(value));
}
function normalizeStringifiedEntries(values) {
	return (values ?? []).map((entry) => normalizeStringifiedOptionalString(entry)).filter((entry) => Boolean(entry));
}
/** Return the trimmed lowercase string or `undefined`. */
function normalizeOptionalLowercaseString(value) {
	return normalizeOptionalString(value)?.toLowerCase();
}
/** Return the trimmed lowercase string or `""`. */
function normalizeLowercaseStringOrEmpty(value) {
	return normalizeOptionalLowercaseString(value) ?? "";
}
/** Coerce a value into a `Record<string, unknown>` or `undefined`. */
function asOptionalObjectRecord(value) {
	return value && typeof value === "object" ? value : void 0;
}
/** Read a string field from a record. */
function readStringField(record, key) {
	const v = record?.[key];
	return typeof v === "string" ? v : void 0;
}
/**
* Normalize filenames into a UTF-8 form that the QQ Bot API accepts reliably.
*
* Decodes percent-escaped names, converts Unicode to NFC, and strips
* ASCII control characters.
*/
function sanitizeFileName(name) {
	if (!name) return name;
	let result = name.trim();
	if (result.includes("%")) try {
		result = decodeURIComponent(result);
	} catch {}
	result = result.normalize("NFC");
	result = result.replace(/\p{Cc}/gu, "");
	return result;
}
//#endregion
export { normalizeStringifiedEntries as a, getPlatformAdapter as c, registerPlatformAdapterFactory as d, normalizeOptionalString as i, hasPlatformAdapter as l, normalizeLowercaseStringOrEmpty as n, readStringField as o, normalizeOptionalLowercaseString as r, sanitizeFileName as s, asOptionalObjectRecord as t, registerPlatformAdapter as u };
