// @openclaw/qqbot/dist/target-parser-DagYRQkP.js
//#region extensions/qqbot/src/engine/messaging/target-parser.ts
/**
* Parse a qqbot target string into a structured delivery target.
*
* Supported formats:
* - `qqbot:c2c:openid` → C2C direct message
* - `qqbot:group:groupid` → Group message
* - `qqbot:channel:channelid` → Channel message
* - `c2c:openid` → C2C (without qqbot: prefix)
* - `group:groupid` → Group (without qqbot: prefix)
* - `channel:channelid` → Channel (without qqbot: prefix)
* - `openid` → C2C (bare openid, default)
*
* @param to - Raw target string.
* @returns Parsed target with type and id.
* @throws {Error} When the target format is invalid.
*/
function parseTarget(to) {
	const id = to.replace(/^qqbot:/i, "");
	if (id.startsWith("c2c:")) {
		const userId = id.slice(4);
		if (!userId) throw new Error(`Invalid c2c target format: ${to} - missing user ID`);
		return {
			type: "c2c",
			id: userId
		};
	}
	if (id.startsWith("group:")) {
		const groupId = id.slice(6);
		if (!groupId) throw new Error(`Invalid group target format: ${to} - missing group ID`);
		return {
			type: "group",
			id: groupId
		};
	}
	if (id.startsWith("channel:")) {
		const channelId = id.slice(8);
		if (!channelId) throw new Error(`Invalid channel target format: ${to} - missing channel ID`);
		return {
			type: "channel",
			id: channelId
		};
	}
	if (!id) throw new Error(`Invalid target format: ${to} - empty ID after removing qqbot: prefix`);
	return {
		type: "c2c",
		id
	};
}
/**
* Normalize a QQ Bot target string into the canonical `qqbot:...` form.
*
* Returns `undefined` when the target does not look like a QQ Bot address.
*/
function normalizeTarget(target) {
	const id = target.replace(/^qqbot:/i, "");
	if (id.startsWith("c2c:") || id.startsWith("group:") || id.startsWith("channel:")) return `qqbot:${id}`;
	if (/^[0-9a-fA-F]{32}$/.test(id)) return `qqbot:c2c:${id}`;
	if (/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/.test(id)) return `qqbot:c2c:${id}`;
}
/**
* Return true when the string looks like a QQ Bot target ID.
*/
function looksLikeQQBotTarget(id) {
	if (/^qqbot:(c2c|group|channel):/i.test(id)) return true;
	if (/^(c2c|group|channel):/i.test(id)) return true;
	if (/^[0-9a-fA-F]{32}$/.test(id)) return true;
	return /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/.test(id);
}
//#endregion
export { normalizeTarget as n, parseTarget as r, looksLikeQQBotTarget as t };
