// @openclaw/qqbot/dist/tools-UJJ-tLHP.js
import { a as resolveQQBotAccount, r as listQQBotAccountIds } from "./config-C1qZbh0K.js";
import { a as formatErrorMessage, n as debugLog, t as debugError } from "./log-DEtcoDWe.js";
import { t as getRequestContext } from "./request-context-Bm7PTBD1.js";
import { truncateUtf16Safe } from "openclaw/plugin-sdk/text-utility-runtime";
import { readProviderTextResponse, readResponseTextLimited } from "openclaw/plugin-sdk/provider-http";
import { fetchWithSsrFGuard } from "openclaw/plugin-sdk/ssrf-runtime";
import { resolveExpiresAtMsFromDurationMs } from "openclaw/plugin-sdk/number-runtime";
import { jsonResult } from "openclaw/plugin-sdk/tool-results";
import { callGatewayTool } from "openclaw/plugin-sdk/agent-harness-runtime";
//#region extensions/qqbot/src/engine/tools/channel-api.ts
/**
* QQ Channel API proxy tool core logic.
* QQ 频道 API 代理工具核心逻辑。
*
* Provides an authenticated HTTP proxy for the QQ Open Platform channel
* APIs. The caller (old tools/channel.ts shell) resolves the access
* token and passes it in; this module handles URL building, path
* validation, fetch, and structured response formatting.
*/
const API_BASE = "https://api.sgroup.qq.com";
const DEFAULT_TIMEOUT_MS = 3e4;
const CHANNEL_API_ERROR_BODY_LIMIT_BYTES = 8 * 1024;
function resolveChannelApiSsrfPolicy(url) {
	return {
		hostnameAllowlist: [new URL(url).hostname],
		allowRfc2544BenchmarkRange: true
	};
}
/**
* JSON Schema for AI tool parameters (used by framework registration).
* AI Tool 参数的 JSON Schema 定义（供框架注册使用）。
*/
const ChannelApiSchema = {
	type: "object",
	properties: {
		method: {
			type: "string",
			description: "HTTP method. Allowed values: GET, POST, PUT, PATCH, DELETE. Use DELETE and other mutating methods only after explicit user intent and target confirmation.",
			enum: [
				"GET",
				"POST",
				"PUT",
				"PATCH",
				"DELETE"
			]
		},
		path: {
			type: "string",
			description: "API path without the host. Replace placeholders with concrete values. Examples: /users/@me/guilds, /guilds/{guild_id}/channels, /channels/{channel_id}."
		},
		body: {
			type: "object",
			description: "JSON request body for POST/PUT/PATCH requests. GET/DELETE usually do not need it. For write requests, include only fields the user explicitly asked to change."
		},
		query: {
			type: "object",
			description: "URL query parameters as key/value pairs appended to the path. For example, { \"limit\": \"100\", \"after\": \"0\" } becomes ?limit=100&after=0.",
			additionalProperties: { type: "string" }
		},
		confirmed: {
			type: "boolean",
			description: "Required true for DELETE requests after the user confirms the exact QQ resource to delete."
		},
		bulkConfirmed: {
			type: "boolean",
			description: "Required true in addition to confirmed for bulk DELETE requests such as deleting all announcements."
		}
	},
	required: ["method", "path"]
};
/**
* Build the full API URL from base + path + query params.
* 拼接 API 基地址 + 路径 + 查询参数。
*/
function buildUrl(path, query) {
	let url = `${API_BASE}${path}`;
	if (query && Object.keys(query).length > 0) {
		const params = new URLSearchParams();
		for (const [key, value] of Object.entries(query)) if (value !== void 0 && value !== null && value !== "") params.set(key, value);
		const qs = params.toString();
		if (qs) url += `?${qs}`;
	}
	return url;
}
/**
* Validate API path format; returns an error string or null if valid.
* 校验 API 路径格式，返回错误描述或 null（合法）。
*/
function validatePath(path) {
	if (!path.startsWith("/")) return "path must start with /";
	if (path.includes("..") || path.includes("//")) return "path must not contain .. or //";
	if (!/^\/[a-zA-Z0-9\-._~:@!$&'()*+,;=/%]+$/.test(path) && path !== "/") return "path contains unsupported characters";
	for (const segment of path.split("/").slice(1)) {
		let decodedSegment;
		try {
			decodedSegment = decodeURIComponent(segment);
		} catch {
			return "path contains invalid percent encoding";
		}
		if (decodedSegment.includes("/") || decodedSegment.includes("\\")) return "path contains encoded path separators";
		if (decodedSegment === "." || decodedSegment === "..") return "path must not contain . or .. segments";
	}
	return null;
}
function decodePathSegments(path) {
	try {
		return path.replace(/\/+$/, "").split("/").slice(1).map((segment) => decodeURIComponent(segment));
	} catch {
		return null;
	}
}
function isBulkAnnouncementDeletePath(path) {
	const segments = decodePathSegments(path);
	return Boolean(segments && segments.length === 4 && segments[0]?.toLowerCase() === "guilds" && segments[2]?.toLowerCase() === "announces" && segments[3]?.toLowerCase() === "all");
}
function validateDeleteConfirmation(params) {
	if (params.method.toUpperCase() !== "DELETE") return null;
	if (!params.confirmed) return "DELETE requests require confirmed=true after the user confirms the exact QQ resource.";
	if (isBulkAnnouncementDeletePath(params.path) && !params.bulkConfirmed) return "Deleting all announcements requires bulkConfirmed=true after a separate bulk-delete confirmation.";
	return null;
}
/**
* Execute a channel API proxy request.
* 执行频道 API 代理请求。
*
* The caller provides the access token; this function handles
* URL building, path validation, HTTP fetch, and structured
* response formatting suitable for AI tool output.
*/
async function executeChannelApi(params, options) {
	if (!params.method) return jsonResult({ error: "method is required" });
	if (!params.path) return jsonResult({ error: "path is required" });
	const method = params.method.toUpperCase();
	if (![
		"GET",
		"POST",
		"PUT",
		"PATCH",
		"DELETE"
	].includes(method)) return jsonResult({ error: `Unsupported HTTP method: ${method}. Allowed values: GET, POST, PUT, PATCH, DELETE` });
	const pathError = validatePath(params.path);
	if (pathError) return jsonResult({ error: pathError });
	const confirmationError = validateDeleteConfirmation({
		...params,
		method
	});
	if (confirmationError) return jsonResult({
		error: confirmationError,
		path: params.path
	});
	if ((method === "GET" || method === "DELETE") && params.body && Object.keys(params.body).length > 0) debugLog(`[qqbot-channel-api] ${method} request with body, body will be ignored`);
	try {
		const url = buildUrl(params.path, params.query);
		const headers = {
			Authorization: `QQBot ${options.accessToken}`,
			"Content-Type": "application/json"
		};
		const controller = new AbortController();
		const timeoutId = setTimeout(() => controller.abort(), DEFAULT_TIMEOUT_MS);
		const fetchOptions = {
			method,
			headers,
			signal: controller.signal
		};
		if (params.body && [
			"POST",
			"PUT",
			"PATCH"
		].includes(method)) fetchOptions.body = JSON.stringify(params.body);
		debugLog(`[qqbot-channel-api] >>> ${method} ${url} (timeout: ${DEFAULT_TIMEOUT_MS}ms)`);
		let res;
		let release;
		try {
			const guarded = await fetchWithSsrFGuard({
				url,
				init: fetchOptions,
				auditContext: "qqbot-channel-api",
				policy: resolveChannelApiSsrfPolicy(url)
			});
			res = guarded.response;
			release = guarded.release;
		} catch (err) {
			clearTimeout(timeoutId);
			if (err instanceof Error && err.name === "AbortError") {
				debugError(`[qqbot-channel-api] <<< Request timeout after ${DEFAULT_TIMEOUT_MS}ms`);
				return jsonResult({
					error: `Request timed out after ${DEFAULT_TIMEOUT_MS}ms`,
					path: params.path
				});
			}
			debugError("[qqbot-channel-api] <<< Network error:", err);
			return jsonResult({
				error: `Network error: ${formatErrorMessage(err)}`,
				path: params.path
			});
		} finally {
			clearTimeout(timeoutId);
		}
		try {
			debugLog(`[qqbot-channel-api] <<< Status: ${res.status} ${res.statusText}`);
			const rawBody = res.ok ? await readProviderTextResponse(res, "QQ channel API response") : await readResponseTextLimited(res, CHANNEL_API_ERROR_BODY_LIMIT_BYTES);
			if (!rawBody || rawBody.trim() === "") {
				if (res.ok) return jsonResult({
					success: true,
					status: res.status,
					path: params.path
				});
				return jsonResult({
					error: `API returned ${res.status} ${res.statusText}`,
					status: res.status,
					path: params.path
				});
			}
			let parsed;
			try {
				parsed = JSON.parse(rawBody);
			} catch {
				parsed = rawBody;
			}
			if (!res.ok) {
				const errMsg = typeof parsed === "object" && parsed && "message" in parsed ? String(parsed.message) : `${res.status} ${res.statusText}`;
				debugError(`[qqbot-channel-api] Error [${method} ${params.path}]: ${errMsg}`);
				return jsonResult({
					error: errMsg,
					status: res.status,
					path: params.path,
					details: parsed
				});
			}
			return jsonResult({
				success: true,
				status: res.status,
				path: params.path,
				data: parsed
			});
		} finally {
			await release?.();
		}
	} catch (err) {
		return jsonResult({
			error: formatErrorMessage(err),
			path: params.path
		});
	}
}
//#endregion
//#region extensions/qqbot/src/bridge/tools/channel.ts
/**
* Register the QQ channel API proxy tool.
*
* The tool acts as an authenticated HTTP proxy for the QQ Open Platform
* channel APIs. Agents learn endpoint details from the skill docs and
* send requests through this proxy.
*/
function registerChannelTool(api) {
	const cfg = api.config;
	if (!cfg) return;
	const accountIds = listQQBotAccountIds(cfg);
	if (accountIds.length === 0) return;
	const firstAccountId = accountIds[0];
	const account = resolveQQBotAccount(cfg, firstAccountId);
	if (!account.appId || !account.clientSecret) return;
	api.registerTool({
		name: "qqbot_channel_api",
		label: "QQBot Channel API",
		description: "Authenticated HTTP proxy for QQ Open Platform channel APIs. Use write and delete endpoints only after explicit user intent; DELETE requires confirmed=true, and bulk deletes require bulkConfirmed=true after confirming the exact target. Common endpoints: list guilds GET /users/@me/guilds | list channels GET /guilds/{guild_id}/channels | get channel GET /channels/{channel_id} | create channel POST /guilds/{guild_id}/channels | list members GET /guilds/{guild_id}/members?after=0&limit=100 | get member GET /guilds/{guild_id}/members/{user_id} | list threads GET /channels/{channel_id}/threads | create thread PUT /channels/{channel_id}/threads | create announce POST /guilds/{guild_id}/announces | create schedule POST /channels/{channel_id}/schedules. See the qqbot-channel skill for full endpoint details.",
		parameters: ChannelApiSchema,
		async execute(_toolCallId, params) {
			const { getAccessToken } = await import("./sender-CjDuU-uz.js").then((n) => n._);
			return executeChannelApi(params, { accessToken: await getAccessToken(account.appId, account.clientSecret) });
		}
	}, { name: "qqbot_channel_api" });
}
//#endregion
//#region extensions/qqbot/src/engine/tools/remind-logic.ts
const QQBOT_DEFAULT_REMINDER_TIMEZONE = "Asia/Shanghai";
/**
* JSON Schema for AI tool parameters (used by framework registration).
* AI Tool 参数的 JSON Schema 定义（供框架注册使用）。
*/
const RemindSchema = {
	type: "object",
	properties: {
		action: {
			type: "string",
			description: "Action type. add=create a reminder only after explicit user request, list=show reminders, remove=delete a reminder by confirmed job ID.",
			enum: [
				"add",
				"list",
				"remove"
			]
		},
		content: {
			type: "string",
			description: "Reminder content, for example \"drink water\" or \"join the meeting\". Required when action=add."
		},
		to: {
			type: "string",
			description: "Optional delivery target. The runtime automatically resolves the current conversation target, so you usually do not need to supply this. Direct-message format: qqbot:c2c:user_openid. Group format: qqbot:group:group_openid."
		},
		time: {
			type: "string",
			description: "Time description. Supported formats:\n1. Relative time, for example \"5m\", \"1h\", \"1h30m\", or \"2d\"\n2. Cron expression, for example \"0 8 * * *\" or \"0 9 * * 1-5\"\nValues containing spaces are treated as cron expressions; everything else is treated as a one-shot relative delay.\nRequired when action=add. Ask for clarification before scheduling if the time is ambiguous."
		},
		timezone: {
			type: "string",
			description: "Optional IANA timezone used for cron reminders. Include it when the user provides or confirms a timezone; if omitted, QQBot preserves its existing default timezone."
		},
		name: {
			type: "string",
			description: "Optional reminder job name. Defaults to the first 20 characters of content."
		},
		jobId: {
			type: "string",
			description: "Job ID to remove. Required when action=remove; fetch it with list first."
		}
	},
	required: ["action"]
};
/**
* Parse a relative time string into milliseconds.
* 解析相对时间字符串为毫秒数。
*
* Supports: "5m", "1h", "1h30m", "2d", "45s", plain number (as minutes).
*
* @returns Milliseconds or null if unparseable.
*/
function parseRelativeTime(timeStr) {
	const s = timeStr.trim().toLowerCase();
	if (/^\d+$/.test(s)) return Number.parseInt(s, 10) * 6e4;
	let totalMs = 0;
	let matched = false;
	let consumed = 0;
	const regex = /(\d+(?:\.\d+)?)\s*(d|h|m|s)\s*/g;
	let match;
	while ((match = regex.exec(s)) !== null) {
		if (match.index !== consumed) return null;
		matched = true;
		consumed = regex.lastIndex;
		const value = Number.parseFloat(match[1]);
		switch (match[2]) {
			case "d":
				totalMs += value * 864e5;
				break;
			case "h":
				totalMs += value * 36e5;
				break;
			case "m":
				totalMs += value * 6e4;
				break;
			case "s":
				totalMs += value * 1e3;
				break;
		}
	}
	return matched && consumed === s.length ? Math.round(totalMs) : null;
}
/**
* Check whether a time string is a cron expression (3–6 space-separated fields).
* 判断时间字符串是否为 cron 表达式。
*/
function isCronExpression(timeStr) {
	const parts = timeStr.trim().split(/\s+/);
	if (parts.length < 3 || parts.length > 6) return false;
	return parts.every((p) => /^[0-9*?/,LW#-]/.test(p));
}
/**
* Generate a cron job name from reminder content (first 20 chars).
* 根据提醒内容生成 cron job 名称。
*/
function generateJobName(content) {
	const trimmed = content.trim();
	return `Reminder: ${trimmed.length > 20 ? `${truncateUtf16Safe(trimmed, 20)}…` : trimmed}`;
}
/** Build the reminder system prompt sent to the AI. */
function buildReminderPrompt(content) {
	return `You are a warm reminder assistant. Please remind the user about: ${content}. Requirements: (1) do not reply with HEARTBEAT_OK (2) do not explain who you are (3) output a direct and caring reminder message (4) you may add a short encouraging line (5) keep it within 2-3 sentences (6) use a small amount of emoji.`;
}
/** Build cron job params for a one-shot delayed reminder. */
function buildOnceJob(params, atMs, to, accountId) {
	const content = params.content;
	return {
		action: "add",
		job: {
			name: params.name || generateJobName(content),
			schedule: {
				kind: "at",
				atMs
			},
			sessionTarget: "isolated",
			wakeMode: "now",
			deleteAfterRun: true,
			payload: {
				kind: "agentTurn",
				message: buildReminderPrompt(content)
			},
			delivery: {
				mode: "announce",
				channel: "qqbot",
				to,
				accountId
			}
		}
	};
}
/** Build cron job params for a recurring cron reminder. */
function buildCronJob(params, to, accountId) {
	const content = params.content;
	const name = params.name || generateJobName(content);
	const tz = params.timezone || QQBOT_DEFAULT_REMINDER_TIMEZONE;
	return {
		action: "add",
		job: {
			name,
			schedule: {
				kind: "cron",
				expr: params.time.trim(),
				tz
			},
			sessionTarget: "isolated",
			wakeMode: "now",
			payload: {
				kind: "agentTurn",
				message: buildReminderPrompt(content)
			},
			delivery: {
				mode: "announce",
				channel: "qqbot",
				to,
				accountId
			}
		}
	};
}
/** Format a delay in milliseconds as a short string (e.g. "5m", "1h30m"). */
function formatDelay(ms) {
	const totalSeconds = Math.round(ms / 1e3);
	if (totalSeconds < 60) return `${totalSeconds}s`;
	const totalMinutes = Math.round(ms / 6e4);
	if (totalMinutes < 60) return `${totalMinutes}m`;
	const hours = Math.floor(totalMinutes / 60);
	const minutes = totalMinutes % 60;
	if (minutes === 0) return `${hours}h`;
	return `${hours}h${minutes}m`;
}
function formatSchedulerError(error) {
	return error instanceof Error ? error.message : String(error);
}
function prepareRemindCronAction(params, ctx = {}) {
	if (params.action === "list") return {
		ok: true,
		action: "list",
		cronAction: { action: "list" }
	};
	if (params.action === "remove") {
		if (!params.jobId) return {
			ok: false,
			error: "jobId is required when action=remove. Use action=list first."
		};
		return {
			ok: true,
			action: "remove",
			cronAction: {
				action: "remove",
				jobId: params.jobId
			}
		};
	}
	if (!params.content) return {
		ok: false,
		error: "content is required when action=add"
	};
	const resolvedTo = params.to || ctx.fallbackTo;
	if (!resolvedTo) return {
		ok: false,
		error: "Unable to determine delivery target for action=add. The reminder can only be scheduled from within an active conversation."
	};
	if (!params.time) return {
		ok: false,
		error: "time is required when action=add"
	};
	const resolvedAccountId = ctx.fallbackAccountId || "default";
	if (isCronExpression(params.time)) return {
		ok: true,
		action: "add",
		cronAction: buildCronJob(params, resolvedTo, resolvedAccountId),
		summary: `⏰ Recurring reminder: "${params.content}" (${params.time}, tz=${params.timezone || QQBOT_DEFAULT_REMINDER_TIMEZONE})`
	};
	const delayMs = parseRelativeTime(params.time);
	if (delayMs == null) return {
		ok: false,
		error: `Could not parse time format: ${params.time}. Use values like 5m, 1h, 1h30m, or a cron expression.`
	};
	if (delayMs < 3e4) return {
		ok: false,
		error: "Reminder delay must be at least 30 seconds"
	};
	const atMs = resolveExpiresAtMsFromDurationMs(delayMs);
	if (atMs === void 0) return {
		ok: false,
		error: "Reminder time is outside the supported Date range"
	};
	return {
		ok: true,
		action: "add",
		cronAction: buildOnceJob(params, atMs, resolvedTo, resolvedAccountId),
		summary: `⏰ Reminder in ${formatDelay(delayMs)}: "${params.content}"`
	};
}
async function executeScheduledRemind(params, ctx, scheduler) {
	const plan = prepareRemindCronAction(params, ctx);
	if (!plan.ok) return jsonResult({ error: plan.error });
	try {
		const cronResult = await scheduler(plan.cronAction);
		return jsonResult({
			ok: true,
			action: plan.action,
			summary: plan.summary,
			cronResult
		});
	} catch (error) {
		return jsonResult({
			error: `Failed to run Gateway cron action: ${formatSchedulerError(error)}`,
			action: plan.action
		});
	}
}
//#endregion
//#region extensions/qqbot/src/bridge/tools/remind.ts
const DEFAULT_GATEWAY_TIMEOUT_MS = 6e4;
function unexpectedCronParams(params) {
	throw new Error(`Unsupported reminder cron action: ${JSON.stringify(params)}`);
}
const defaultDeps = { callCron: async (params) => {
	switch (params.action) {
		case "list": return await callGatewayTool("cron.list", { timeoutMs: DEFAULT_GATEWAY_TIMEOUT_MS }, {});
		case "remove": return await callGatewayTool("cron.remove", { timeoutMs: DEFAULT_GATEWAY_TIMEOUT_MS }, { jobId: params.jobId });
		case "add": return await callGatewayTool("cron.add", { timeoutMs: DEFAULT_GATEWAY_TIMEOUT_MS }, { job: params.job });
	}
	return unexpectedCronParams(params);
} };
function createRemindTool(toolContext = {}, deps = defaultDeps) {
	return {
		name: "qqbot_remind",
		label: "QQBot Reminder",
		description: "Create, list, and remove QQ reminders. Use only for explicit user requests, and ask when reminder content, schedule, or timezone is ambiguous. This tool schedules Gateway cron jobs directly; do not call the cron tool after it succeeds.\nCreate: action=add, content=message, time=schedule (to is optional, resolved automatically from the current conversation)\nList: action=list\nRemove: action=remove, jobId=job id from list\nTime examples: \"5m\", \"1h\", \"0 8 * * *\"; include timezone for recurring cron reminders when known.",
		parameters: RemindSchema,
		async execute(_toolCallId, params) {
			const ctx = getRequestContext();
			return await executeScheduledRemind(params, {
				fallbackTo: ctx?.target ?? toolContext.deliveryContext?.to,
				fallbackAccountId: ctx?.accountId ?? toolContext.deliveryContext?.accountId
			}, deps.callCron);
		}
	};
}
function registerRemindTool(api) {
	api.registerTool((ctx) => createRemindTool(ctx), { name: "qqbot_remind" });
}
//#endregion
//#region extensions/qqbot/src/bridge/tools/index.ts
function registerQQBotTools(api) {
	registerChannelTool(api);
	registerRemindTool(api);
}
//#endregion
export { registerRemindTool as n, registerChannelTool as r, registerQQBotTools as t };
